---
title: "DADA2 pipeline of the Firefungi 2021 dataset sampling"
author: "Amara Santiesteban Serrano"
date: " r format(Sys.time(), '%d %B, %Y') "
output: 
  html_document:
    df_print: paged
    pdf_document: default
    word_document: default
  editor_options:
    chunk_output_type: inline
---
############### SET THE CURRENT FOLDER AS WORKING DIRECTORY

```{r Packages}
#install.packages("pacman")
pacman::p_load("ggplot2","tidyr","dplyr","tidyverse","readxl","pastecs","lattice","forcats","corrgram","corrplot","HH","effects","car","multcompView","lme4", "devtools","emmeans","factoextra","ggfortify","RColorBrewer","MuMIn","scales","multifunc", "ggthemes", "knitr", "BiocManager", "phyloseq", "phangorn", "DECIPHER", "biomformat", "rbiom", "R.utils", "Biostrings", "ShortRead")
```

The *DADA (Divisive Amplicon Denoising Algorithm)* is a model-based approach for correcting amplicon errors without constructing OTUs. It identifies fine-scale variation.*The output of DADA2 is always ASV* --> trabajar con ASVs tiene sentido principalmente cuando queremos "afinar" al nivel de cepas
--> *DADA2* is the open-source R package that allows researchers to accurately reconstruct amplicon-sequenced communities at the highest resolution.

################################  CONCEPT  NOTE     #######################################################
*OTU* = Operational Taxonomic Unit (se basa en agrupar sequencias entorno a un umbral basado en similitud y definido por un marcador molecular o barcode. Va a depender de la resolución taxonómica que estemos buscando, ej.: 95% para género y 97% para especie)
*ASV* = Amplicon sequence variant (¡representan variantes genéticas únicas! Incluso distinguen entre secuencias que varian entre ellas por un solo nucleótido). 
###########################################################################################################

```{r DADA2 install}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("dada2", version = "3.20")

library(dada2)
```

-*OUR DATA*-
· The pool was sequenced in a fraction of a NovaSeq PE250 flow cell (Illumina) aiming for a total output of 10 gigabases.
  --> 10 gigabases = 10^10 bp
· We have *Illumina paired-end* data (DADA2 was created specifically for this kind of data even though it works with other sequencing techniques). Our amplifying platform is *NovaSeq PE250*: this means PAIR END of 250 bp --> 250 forward and 250 reverse.
· The *amplicon* we used is the ITS (primers used were ITS1f and ITS2) --> *ITS is an hyper variable region* from 80 pb to 800 pb... (*SIZE POLYMORFISM*). In ~AllGenetics~ they said they amplified an ITS region of around 265 pb. 
· Finished libraries were pooled in equimolar amounts according to the results of a Qubit dsDNA HS Assay (Thermo Fisher Scientific) quantification.
· ~AllGenetics~ removed all adapter dimers from the Raw data using #Cutadapt v3.5 

  ######################### How can we know this sequencing depth is enough for us? ###########################
  
  We asked for 10^10 pb --> Illumina Novaseq PE250 aims for reads of 250 pb so 10^10/250 = 4x10^7 --> but these reads are paired-end so   (4x10^7)/2 = 2x10^7 --> if we have 80 samples, theoretically we would have (2x10^7)/80 = 250000 pairs per sample
  
  If we wanted to sequence a whole genome with this sequencing depth: 
  Size of a whole genome (for example) = 35x10^8 pb --> (10^10)/(35x10^8) = 2.85 reads, so this means the whole genome would only be     completely sequenced 2.85 times which is totally insufficient to have a security on the sequencing
  
  #############################################################################################################
  
```{r Identification of primers}

# 1) Unzip fastq files
tgz_file <- "Raw_data.tgz"
untar(tgz_file) ##unzip

# 2) Select the sequences of the project you want to work with:
# Define the path to the unzipped "Raw_data" folder
path_raw_data <- "Raw_data"  
# List the files in the "Raw_data" folder that contain the word "FIREFUNGI"

#firefungi_files <- list.files(path_raw_data, pattern = "FIREFUNGI", full.names = TRUE)
root_2021_files <- list.files(path_raw_data, pattern = "MICORRIZAS", full.names = TRUE)

# Define the destination folder for the "FIREFUNGI" files
destination_folder <-("Raw_data_firefungi_roots") 
# Create the destination folder if it doesn't exist
if (!dir.exists(destination_folder)) {
  dir.create(destination_folder, recursive = TRUE)}
# Move the files to the destination folder
file.copy(root_2021_files, destination_folder, overwrite = TRUE)

#path_firefungi
path_roots <- "Raw_data_firefungi_roots" # CHANGE ME to the directory containing the fastq files after unzipping.
list.files(path_roots)

# Read in the names of the fastq files, and perform some string manipulation to get matched lists of the forward and reverse fastq files. 
fnFs <- sort(list.files(path_firefungi, pattern="_R1_001.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path_firefungi, pattern="_R2_001.fastq.gz", full.names = TRUE))

fnFs_roots <- sort(list.files(path_roots, pattern="_R1_001.fastq.gz", full.names = TRUE))
fnRs_roots <- sort(list.files(path_roots, pattern="_R2_001.fastq.gz", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)

###################### To know the total number of raw sequences #######################################
# Function to count reads in a gzipped fastq file
count_reads <- function(fastq_file) {
  n <- length(readLines(gzfile(fastq_file))) / 4
  return(n)
}

# Count reads for all forward and reverse reads
read_counts_F <- sapply(fnFs_roots, count_reads)
read_counts_R <- sapply(fnRs_roots, count_reads)

# Total reads per direction and in total
total_reads_F <- sum(read_counts_F)
total_reads_R <- sum(read_counts_R)

# Print summary
cat("Total forward reads:", total_reads_F, "\n")
cat("Total reverse reads:", total_reads_R, "\n")
cat("Total raw paired reads:", min(total_reads_F, total_reads_R), "\n") # conservative estimate
########################################################################################################

# Identify primers:
FWD <- "CTTGGTCATTTAGAGGAAGTAA"  
REV <- "GCTGCGTTCTTCATCGATGC"

#Verification of the presence and orientation of these primers in the data: 
allOrients <- function(primer) {
    # Create all orientations of the input sequence
    require(Biostrings)
    dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors   
    orients <- c(Forward = dna, Complement = Biostrings::complement(dna), Reverse = Biostrings::reverse(dna),
        RevComp = Biostrings::reverseComplement(dna))
    return(sapply(orients, toString))  # Convert back to character vector
}
FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients

#The presence of ambiguous bases (Ns) in the sequencing reads makes accurate mapping of short primer sequences difficult. Next we are going to “pre-filter” the sequences just to remove those with Ns, but perform no other filtering.
fnFs_roots.filtN <- file.path(path_roots, "filtN", basename(fnFs_roots)) # Put N-filtered files in filtN/ subdirectory
fnRs_roots.filtN <- file.path(path_roots, "filtN", basename(fnRs_roots))
#Filtered
Filtered_roots <- dada2::filterAndTrim(fnFs_roots, fnFs_roots.filtN, fnRs_roots, fnRs_roots.filtN, maxN = 0, multithread = FALSE) #multithread = TRUE --> this is only for UBUNTU or LINUX so the computer would run it in different threads
Filtered_roots
str(Filtered_roots)
#write.table(Filtered, "Filtered_Ns.txt") # .txt file with initial input sequences and final output sequences after Ns filtering (results will only show number of reads of the Forward sequence, so we consider it as the number of reads of pair sequences, Fwd-Rvs).

#The length variation of the ITS region has significant consequences for the filtering and trimming steps of the standard DADA2 workflow: one of them is that primers can be found within the reads in some cases. We are now ready to count the number of times the primers appear in the forward and reverse read, while considering all possible primer orientations. Identifying and counting the primers on one set of paired end FASTQ files is sufficient, assuming all the files were created using the same library preparation, so we’ll just process the first sample:
primerHits <- function(primer, fn) {
    # Counts number of reads in which the primer is found
    nhits <- Biostrings::vcountPattern(primer, ShortRead::sread(ShortRead::readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0))
}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs_roots.filtN[[1]]), FWD.ReverseReads = sapply(FWD.orients,
    primerHits, fn = fnRs_roots.filtN[[1]]), REV.ForwardReads = sapply(REV.orients, primerHits,
    fn = fnFs_roots.filtN[[1]]), REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs_roots.filtN[[1]]))
```

```{r Primers removal - Cutadapt}

## Cutadapt was developed for Linux so we need to do some adjustments to implement it in Windows: 
## 1) Download the Cutadapt .exe from the web: https://github.com/marcelm/cutadapt/releases
## 2) Put the .exe in the folder you need 
library(ShortRead); packageVersion("ShortRead")
# Set up pathway to cutadapt (primer trimming tool) and test
cutadapt <- ("cutadapt.exe")
system2(cutadapt, args = "--version") # 4.0

#Remove primers with cutadapt and assess the output
# Create directory to hold the output from cutadapt
#trimmed.d <- ("output_cutadapt")
trimmed.d_root <- ("output_cutadapt_root")
# Create the destination folder if it doesn't exist
if (!dir.exists(trimmed.d_root)) {
  dir.create(trimmed.d_root, recursive = TRUE)}

fnFs.cut_root <- file.path(trimmed.d_root, basename(fnFs_roots))
fnRs.cut_root <- file.path(trimmed.d_root, basename(fnRs_roots))

# Save the reverse complements of the primers to variables
FWD.RC <- dada2:::rc(FWD)
REV.RC <- dada2:::rc(REV)

##  Create the cutadapt flags ##
# Trim FWD and the reverse-complement of REV off of R1 (forward reads)
R1.flags <- paste("-g", FWD, "-a", REV.RC, "--minimum-length 50")
# Trim REV and the reverse-complement of FWD off of R2 (reverse reads)
R2.flags <- paste("-G", REV, "-A", FWD.RC, "--minimum-length 50")

# Run Cutadapt
for (i in seq_along(fnFs_roots)) {
  system2(cutadapt, args = c(R1.flags, R2.flags, "-n", 2, #"--revcomp", # -n 2 required to remove FWD and REV from reads
                             "-o", fnFs.cut_root[i], "-p", fnRs.cut_root[i], # output files
                             fnFs_roots.filtN[i], fnRs_roots.filtN[i])) # input files
}

 # As a sanity check, we will check for primers in the first cutadapt-ed sample:
## should all be zero!
rbind(FWD.ForwardReads_root = sapply(FWD.orients, primerHits, fn = fnFs.cut_root[[18]]),
      FWD.ReverseReads_root = sapply(FWD.orients, primerHits, fn = fnRs.cut_root[[18]]),
      REV.ForwardReads_root = sapply(REV.orients, primerHits, fn = fnFs.cut_root[[18]]),
      REV.ReverseReads_root = sapply(REV.orients, primerHits, fn = fnRs.cut_root[[18]]))
#Success! Primers are no longer detected in the cutadapted reads.

#The primer-free sequence files are now ready to be analyzed through the DADA2 pipeline. Similar to the earlier steps of reading in FASTQ files, we read in the names of the cutadapt-ed FASTQ files and applying some string manipulation to get the matched lists of forward and reverse fastq files.

# Forward and reverse fastq filenames have the format:
cutFs_root <- sort(list.files(trimmed.d_root, pattern = "_R1_001.fastq.gz", full.names = TRUE))
cutRs_root <- sort(list.files(trimmed.d_root, pattern = "_R2_001.fastq.gz", full.names = TRUE))

# Extract sample names, assuming filenames have format:
get.sample.name <- function(fname) strsplit(basename(fname), "_")[[1]][1]
sample.names <- unname(sapply(cutFs, get.sample.name))
head(sample.names)
```

```{r DADA2}
#Visualization --> the green line is the average quality score
#dada2::plotQualityProfile(cutFs[c(1:2)]) #Change these numbers to see different samples
#Qp + theme(
#   strip.text = element_text(size = 4.5))  
#ggsave("qualityplotsF.png")

#Filter and trim
# Place filtered files in filtered/ subdirectory (TO PLACE OUTPUT FROM NEXT STEP)
filtFs_root <- file.path(trimmed.d_root, "filtered", basename(cutFs_root))
filtRs_root <- file.path(trimmed.d_root, "filtered", basename(cutRs_root))

                  ################ The next steps can take several minutes ##################

out_roots <- dada2::filterAndTrim(cutFs_root, filtFs_root, cutRs_root, filtRs_root, maxN = 0, maxEE = c(2, 2), truncQ = 2,
                     #cutFs = first input (forward); filtFs = were do we want the output
                     #cutRs = second input (reverse); filtRs = were do we want the output
                     ######### we leave truncLen out when working with ITS since it is an hyper variable region
                     #maxN = reads with ambiguous pairs allowed
                     #maxEE = maximum amount of estimated errors
       minLen = 50, rm.phix = TRUE, compress = TRUE, multithread = FALSE)  # on windows, set multithread = FALSE
head(out_roots)
#write.table(out, "out.txt") 

# Total reads after filtering
total_filtered_root_reads <- sum(out_roots[, "reads.out"])
cat("Total reads after filtering:", total_filtered_root_reads, "\n")

#Error rates for each possible nucleotide transition --> it creates an error model that will be used downstream in the hardcore of the DADA2 algorithm
errF_root <- dada2::learnErrors(filtFs_root, multithread=FALSE)
errR_root <- dada2::learnErrors(filtRs_root, multithread=FALSE)
dada2::plotErrors(errF_root, nominalQ=TRUE)
dada2::plotErrors(errR_root, nominalQ=TRUE)

#Sample Inference --> the hardcore of the DADA2 algorithm
dadaFs_root <- dada2::dada(filtFs_root, err = errF_root, multithread = FALSE)
dadaRs_root <- dada2::dada(filtRs_root, err = errR_root, multithread = FALSE)

#Merge pair reads:
#We now merge the forward and reverse reads together to obtain the full denoised sequences. Merging is performed by aligning the denoised forward reads with the reverse-complement of the corresponding denoised reverse reads, and then constructing the merged “contig” sequences. By default, merged sequences are only output if the forward and reverse reads overlap 
mergers_root <- dada2::mergePairs(dadaFs_root, filtFs_root, dadaRs_root, filtRs_root, verbose=TRUE)
# Inspect the merger data.frame from the first sample
View(mergers_root[[1]])

#Construct sequence table:
seqtab_root <- dada2::makeSequenceTable(mergers_root)
dim(seqtab_root)
# Inspect distribution of sequence lengths
seq_root_lenght <- table(nchar(dada2::getSequences(seqtab_root))) #upper numbers = seq length, numbers below = nº of seqs with that length

#Remove chimeras:
seqtab_root.nochim <- dada2::removeBimeraDenovo(seqtab_root, method="consensus", multithread=FALSE, verbose=TRUE)
#write.csv(seqtab.nochim, file = "seqtab.nochim.csv", row.names = TRUE)
dim(seqtab_root.nochim)
sum(seqtab_root.nochim)/sum(seqtab_root)

# Total reads after chimera removal
total_reads_after <- sum(seqtab_root.nochim)
cat("Total reads after chimera removal:", total_reads_after, "\n")
#Number of ASV after chimera removal
n_asvs <- ncol(seqtab_root.nochim)
cat("total_ASVs", n_asvs, "\n")
# Sum the total reads per sample
sample_sums <- rowSums(seqtab_root.nochim)
# Get the average number of sequences per sample
average_sequences <- mean(sample_sums)
# Print the result
cat("Average number of sequences per sample:", average_sequences, "\n")
```

```{r Tracking}
# Function to get number of reads per sample from mergers
getN <- function(x) sum(dada2::getUniques(x))

# Build tracking table
track_dada2_root <- data.frame(
  input = out_roots[, "reads.in"],
  filtered = out_roots[, "reads.out"],
  denoisedF = sapply(dadaFs_root, getN),
  denoisedR = sapply(dadaRs_root, getN),
  merged = sapply(mergers_root, getN),
  nonchim = rowSums(seqtab_root.nochim)
)
#write.csv(track, "track_dada2_root.csv")

rownames(track) <- names(out_roots)
head(track)
```

```{r Assign taxonomy}
#The assignTaxonomy function takes as input a set of sequences to be classified, and a training set of reference sequences with known taxonomy, and outputs taxonomic assignments with at least minBoot bootstrap confidence.
#The database we are using in this case is **UNITE 2023**: https://unite.ut.ee/repository.php --> this is critical because depending on which database we use and even the version, it can drastically change our results.

# Assign taxonomy                 ###### ATENTION: HIGH COMPUTATIONAL EFFORT ######
                                   ### MORE THAN 1 DAY OF COMPUTING IN WINDOWS ###

taxa_root <- dada2::assignTaxonomy(seqtab_root.nochim, "sh_general_release_dynamic_s_all_25.07.2023.fasta", minBoot = 0, multithread=FALSE, outputBootstraps = TRUE) #My default DADA2 considers a 50% Boostrap threshold, but we wanted to try with a higher one (minBoot = 80%))
View(taxa_root$tax)

##########################################################################################################################
#We could also set the minBoot to 0 --> this will give us a boostrap value per ASV so we can interpret and filter ourselves
##########################################################################################################################

taxa_final2 <- cbind(taxa_root[[1]], taxa_root[[2]])
#write.csv(taxa_final, "taxa_final_BTS.csv")

####################################################

#write.csv(cbind(t(seqtab_root.nochim), taxa_final), file="total_matrix.csv")

#Let’s inspect the taxonomic assignments:
taxa.print_root <- taxa_root # Removing sequence rownames for display only
rownames(taxa.print_root) <- NULL
head(taxa.print_root)

#Percentage of ASVs assigned at each taxonomic level
# Total number of ASVs
total_asvs <- nrow(taxa_root)
# Percentage of assigned ASVs at each taxonomic level
percent_assigned <- colSums(!is.na(taxa_root)) / total_asvs * 100
# Round for readability
round(percent_assigned, 1)
```

*Bootstrap* --> para asignación taxonómica, independientemente de los organismos con los que estemos tratando. Coge un "trocito" de secuencia (subsecuencia o K-mer) de un determinado número de pares de bases (8 pb en este caso) que se van a repetir muchas veces. Estas secuencias se van a comparar con la base de datos en cuestión (UNITE en este caso) y si más de un 80% de las subsecuencias se pegan a un mismo taxón, el algoritmo asignará la secuencia madre a este taxón.

Nuestro output en este punto es todo lo que necesitamos para analizar nuestro set de datos. No obstante, trabajar con toda la secuencia de nucleótidos es más difícil además de más costoso a nivel del procesado de datos. Por esta razón, vamos a sustituir las largas secuencia por un nombre tal como *"ASV_"*:

```{r replace sequences with ASV}

output_folder <- ("Raw_data_firefungi_roots/dada2_output/")
# Create the destination folder if it doesn't exist
if (!dir.exists(output_folder)) {
  dir.create(output_folder, recursive = TRUE)}

#transposed table and replace sequences with ASV (ASV_table.csv)
#Extracted sequences and ASVs to make a fasta file (asv_OTU.fasta)

asv_seqs <- colnames(seqtab_root.nochim)
asv_headers <- vector(dim(seqtab_root.nochim)[2], mode="character")

#fasta header as ASV_
for (i in 1:dim(seqtab_root.nochim)[2]) {
  asv_headers[i] <- paste(">ASV", i, sep="_")
}

#ASV sequences
asv_fasta <- c(rbind(asv_headers, asv_seqs))
View(asv_fasta)

#ASV abundance
asv_table <- t(seqtab_root.nochim)
#head(seqtab_root.nochim)
row.names(asv_table) <- sub(">", "", asv_headers)

#ASV taxonomy
asv_tax <- as.data.frame(taxa_root$tax)
row.names(asv_tax) <- sub(">", "", asv_headers)

#merging abundance and tax table
OTU_TAX_table <- merge(asv_table, asv_tax, by=0)
OTU_TAX_table

#defining output paths within the new folder
asv_fasta_file <- file.path(output_folder, "asv_OTU.fasta")
asv_table_file <- file.path(output_folder, "asv_table.csv")
asv_tax_file <- paste0(output_folder, "asv_tax.csv")
OTU_TAX_table_file <- paste0(output_folder, "OTU_TAX_table.csv")

#writing out output files
write.table(asv_fasta, asv_fasta_file, sep=",", quote=F, col.names=FALSE,row.names = FALSE)
#write.table(asv_table, asv_table_file, sep=",", quote=F, col.names = NA)
#write.table(asv_tax, asv_tax_file, sep=",", quote=F, col.names=NA)
#write.table(OTU_TAX_table, OTU_TAX_table_file, sep=",", quote=F, col.names=NA)
```

DADA2 tiene sentido para afinar a nivel de ASV, es una máquina súper potente. Pero para trabajar con OTUs podríamos usar algoritmos menos profundos como VSEARCH. Pero usamos DADA2 porque está de moda...
