---
title: "Phyloseq analysis of Navalacruz 2021 dataset sampling"
author: "Amara Santiesteban Serrano"
date: " r format(Sys.time(), '%d %B, %Y') "
output: 
  html_document:
    df_print: paged
    pdf_document: default
    word_document: default
  editor_options:
    chunk_output_type: inline
---

*Primero procesar las secuencias en DADA2 (1Âº) y en LULU (2Âº)*

```{r Packages}
#install.packages("pacman")
pacman::p_load("ggplot2","tidyr","dplyr","tidyverse","readxl","pastecs","lattice","forcats","corrgram","corrplot","HH","effects","car","multcompView","lme4", "devtools","emmeans","factoextra","ggfortify","RColorBrewer","MuMIn","scales","multifunc", "ggthemes", "knitr", "BiocManager", "phyloseq", "phangorn", "DECIPHER", "biomformat", "rbiom", "plyr", "indicspecies", "networkD3", "htmlwidgets", "webshot2", "vegan", "lme4", "car", "plyr", "ggpubr", "ggsignif")
```

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("microbiome")
```

```{r Colors}
cols <- c("#41b6c4","#fd8d3c","#cb181d","#800026")
```

```{r data required}
#########################
#OTU abundance: *matrix*            class(clustered_asv_table)
#Taxonomy: *matrix*                 class(clustered_taxa)
#Metadata: *data.frame*
#FunGuild: *matrix*
########################

#We need the taxa table and the curated asv table to have the same number of entries
#We need the meta_data and the curated asv table to have the exact same samples names

#####OTU abundance:
colnames(clustered_asv_table) <- sapply(colnames(clustered_asv_table), function(name) {
  parts <- strsplit(name, "-FIREFUNGI")[[1]]
  return(paste0(parts[1], "-FIREFUNGI"))
})
#clustered_asv_table[] <- lapply(clustered_asv_table, as.numeric)

####Taxonomy: 
class(clustered_tax_table)
clustered_tax_table <- as.matrix(clustered_tax_table)

####Metadata
meta_data <- as.data.frame(Firefungi_soil_dataset_2021)
rownames(meta_data) <- paste0(1:80, "-FIREFUNGI")

####FunGuild
#FG_matrix <- select(FG, confidenceRanking, trophicMode, guild, growthForm)
#FG_matrix <- as(as.matrix(FG_matrix),"matrix")
```

The *phyloseq package* is a tool to import, store, analyze, and graphically display complex phylogenetic sequencing data that has already been clustered into Operational Taxonomic Units (OTUs), especially when there is associated sample data, phylogenetic tree, and/or taxonomic assignment of the OTUs. This package leverages many of the tools available in R for ecology and phylogenetic analysis (vegan, ade4, ape, picante), while also using advanced/flexible graphic systems (ggplot2) to easily produce publication-quality graphics of complex phylogenetic data. 

```{r Phyloseq object}
ps <- phyloseq(otu_table(clustered_asv_table, taxa_are_rows=TRUE), 
               sample_data(meta_data, sample_names(sample.names)),
               tax_table(clustered_taxa))
ps
```

```{r}
# We make sure that we do not have OTUs that are not present in any of the samples:
ps0 <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
ps0
#We also did a filtering of ASV with less than 5 reads but the result is the same (probably due to LULU & postclustering)
```

*ADVICE*: check some sequences (from the 'asv_OTU.fasta' generated in DADA2, they can be accessed through Notepad++) in BLAST: https://blast.ncbi.nlm.nih.gov/Blast.cgi# --> copy and paste the whole sequence in the Entry query sequence box, and click on 'BLAST' at the end. This is a double check to see if our reference database (UNITE in this case) has assigned correctly the taxons.

```{r Alpha Diversity}
# Observed = Richness: total number of OTU observed in a sample (it does not consider abundance or eveness)
# Chao1: non-parametric estimator of the minimum species richness. The number of species in a community, accounting for the number of rare species that are likely to be missed
# Shannon: It measures the diversity of a community by considering both species richness and evenness. Higher values indicate greater diversity, with more even distribution of species abundances.

alpha <- microbiome::alpha(ps, index = c("Observed", "Shannon", "Chao1"))

alpha <- alpha %>% 
  tibble::rownames_to_column(var = "SampleID")

metadata_df <- meta_data %>% 
  tibble::rownames_to_column(var = "SampleID")

alpha <- left_join(alpha, metadata_df, by = "SampleID")
```

```{r Alpha Diversity Models}
alpha$TRAT2 <- ordered(alpha$TRAT2, levels=c("UB", "LS", "HS1", "HS2"))
alpha$DEPTH <- ordered(alpha$DEPTH, levels=c("topsoil", "subsoil"))

### Visually assessment of data
var <- alpha$diversity_shannon

hist(var) 

alpha %>%
  ggplot(aes(x=TRAT2, y=var, fill=DEPTH)) +
  geom_boxplot(outlier.color = "red",
    outlier.shape = 8) +
  scale_fill_brewer(palette="Dark2") +
  stat_summary(fun= mean, geom="point", shape=20, size=2, colour="black") + 
  facet_wrap(~ DEPTH, scales="free_x") 

#################### WHOLE ##################################
#### FIRST!! We define the Contrast table we want to work with
options(contrasts=c(factor="contr.sum", ordered="contr.poly"))
##############################################################
mixed.var <- lmer(var ~ TRAT2*DEPTH + (1|SITE), data = alpha)

par(mfrow=c(1,3))
qqnorm(resid(mixed.var)); qqline(resid(mixed.var))
hist(resid(mixed.var))
plot(fitted(mixed.var), resid(mixed.var)); abline(h=0)
par(mfrow=c(1,1))

car::Anova(mixed.var, test.statistic="F", type=3)

multcomp::cld(emmeans(mixed.var, ~TRAT2, type="response"), Letters= letters) #Tuckey's post-hoc test
```

-No significant differences found across *DEPTHS*
-No significant differences in *Shannon diversity* across treatments
-Significant differences foun in *observed and Chao1 richness*: UB > LS > HS1 > HS2

```{r Alpha Diversity Plot}
# Specify the desired order of levels for TRAT2
TRAT2_order <- c("UB", "LS", "HS1", "HS2")  
# Reorder the levels of TRAT2
sample_data(ps)$TRAT2 <- factor(sample_data(ps)$TRAT2, levels = TRAT2_order)

plot_richness(ps, color = "TRAT2", x = "TRAT2", 
              measures = c("Observed", "Shannon", "Chao1")) + 
  geom_boxplot(aes(fill = TRAT2), alpha=.7) + 
  scale_color_manual(values = c("#1d91c0","#feb24c","#cb181d","#800026")) + 
  scale_fill_manual(values = c("#1d91c0","#feb24c","#cb181d","#800026")) +
  theme(
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(),
    strip.background = element_blank(),  # Remove strip background
    axis.text.x = element_text(angle = 0.1, hjust = 1),  # Rotate x-axis labels
    panel.background = element_blank(),  # Remove panel background color
    plot.background = element_blank(),   # Remove plot background color
    legend.position = "none"             # Remove legend
  )
#ggsave("alpha_div.png")
```

```{r Quality analysis}
# Define prevalence of each taxa (in how many samples did each taxa appear at least once)
# Compute prevalence (number of samples where an OTU is present)
prevdf <- apply(X = otu_table(ps),
                MARGIN = ifelse(taxa_are_rows(ps), yes = 1, no = 2),
                FUN = function(x) sum(x > 0))

prevdf <- data.frame(Prevalence = prevdf, row.names = taxa_names(ps))

# Add total abundance and taxonomy information
prevdf <- data.frame(Prevalence = prevdf$Prevalence,
                     TotalAbundance = phyloseq::taxa_sums(ps),
                     as.data.frame(tax_table(ps)))

#With the following table we can see how well represented is each "Phylum" (or Kingdom, etc...)
dfprev <- plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
```

```{r Filtering: unlikely taxa}
#Based on the results above, we could decide to filter some Phylum out (the ones with less representation). We can also get rid of the Kingdoms not corresponding to fungi:
plot_bar(ps, fill = "Kingdom") + 
geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack")

ps1 = phyloseq::subset_taxa(ps, Kingdom=="k__Fungi")
ps1

#plot_bar(ps1, fill = "Kingdom") + 
#geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack")
```

```{r Visualization}
# Define prevalence threshold as 5% of total samples 
prevalenceThreshold = 0.05 * phyloseq::nsamples(ps1)
prevalenceThreshold

#This plot will help us to identify which taxa are more abundant and prevalent in the dataset.
ggplot(prevdf, aes(TotalAbundance, Prevalence, color = Phylum)) + 
  geom_hline(yintercept = prevalenceThreshold, alpha = 0.5, linetype = 2) + 
  geom_point(size = 2, alpha = 0.7) + 
  scale_y_log10() + 
  scale_x_log10() + 
  xlab("Total Abundance") + 
  facet_wrap(~Phylum) +
  theme(legend.position = "none")
#ggsave("prevalence.png")

#This will show us the distribution of read counts per sample:
sample_sum_df <- data.frame(sum = sample_sums(ps1))
ggplot(sample_sum_df, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "grey", binwidth = 2500) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  theme(axis.title.y = element_blank()) 
```

```{r Most abundant 5 phyla}
phylum.sum = tapply(phyloseq::taxa_sums(ps1), tax_table(ps1)[, "Phylum"], sum, na.rm=TRUE)
top5phyla = names(sort(phylum.sum, TRUE))[1:5]
ps2 = phyloseq::prune_taxa((tax_table(ps1)[, "Phylum"] %in% top5phyla), ps1)
ps2

plot_bar(ps2, "Phylum", fill = "Phylum", facet_grid = ~TRAT2)
```

```{r Ascomycota/Basidiomycota}
# Filter to include only specified phyla
filterPhylaAB <- c("p__Ascomycota", "p__Basidiomycota")
ps4 = phyloseq::subset_taxa(ps2, Phylum %in% filterPhylaAB)
# Normalize the data before filtering
p <- transform_sample_counts(ps2, function(x) { x / sum(x) } * 100)
# Filter the normalized data
ps4 <- phyloseq::subset_taxa(p, Phylum %in% filterPhylaAB)
# Melt the filtered and normalized data
ps4melted <- psmelt(ps4)

####### Summarize the data to calculate mean and SE 
summary_data <- ps4melted %>%
  dplyr::group_by(Phylum, DEPTH, TRAT2) %>%
  dplyr::summarise(
    median_abundance = median(Abundance),
    mean_abundance = mean(Abundance),
    sd_abundance = sd(Abundance),
    se_abundance = sd(Abundance) / sqrt(n()))  # Calculate standard error

ps4melted <- ps4melted %>%
  left_join(summary_data, by = c("Phylum", "DEPTH", "TRAT2"))

# Set factors for DEPTH and TRAT2
ps4melted$DEPTH <- factor(ps4melted$DEPTH, levels = c("topsoil", "subsoil"))
ps4melted$TRAT2 <- factor(ps4melted$TRAT2, levels = c("UB", "LS", "HS1", "HS2"))

cols_2 <- c("#fed976","#fd8d3c")

# Create the plot
ggplot(ps4melted, aes(x = Phylum, y = mean_abundance)) +
  geom_bar(aes(color = Phylum, fill = Phylum), stat = "identity", position = "dodge") + 
  geom_errorbar(aes(ymin = mean_abundance - se_abundance, ymax = mean_abundance + se_abundance), width = 0.2, position = position_dodge(0.9)) +
  scale_color_manual(values = cols_2) +  # Use a Brewer palette
  scale_fill_manual(values = cols_2) +
  facet_grid(DEPTH ~ TRAT2) +
  theme(
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black"),
    axis.text.x = element_blank(),
    axis.ticks = element_line(color = "black"),
    strip.background = element_blank(),  # Remove strip background
    legend.position = "none"
  )
#ggsave("AscoBasidio.png")


```

```{r Ascomycota/Basidiomycota Models}

# Fit a Linear Mixed-Effects Model
# Here, random effect is Sample (or any grouping variable) and fixed effects are Phylum, DEPTH, and TRAT2
# Standardize the continuous variables
ps4melted$mean_abundance <- scale(ps4melted$mean_abundance)


lmm_results <- lmer(mean_abundance ~ Phylum * DEPTH * TRAT2 + (1|Sample), data = ps4melted)

par(mfrow=c(1,3))
qqnorm(resid(lmm_results)); qqline(resid(lmm_results))
hist(resid(lmm_results))
plot(fitted(lmm_results), resid(lmm_results)); abline(h=0)
par(mfrow=c(1,1))
```

```{r Global summary of the phyloseq object}
microbiome::summarize_phyloseq(ps2) #Results total number of reads and average number of reads by sample

global_df <- psmelt(ps2)
head(global_df)
#write.csv(global_df, "global_df.csv")
```

*Beta diversity* is going to account for the *turnover* across sites, treatments...

```{r Beta diversity}
# Based on distances --> we work with the Hellinger transformed data
GP.ord <- ordinate(ps2, "NMDS", "bray")
p1 = plot_ordination(ps2, GP.ord, type="taxa", color="Phylum", title="taxa")
print(p1)

p1 + facet_wrap(~Phylum, 3)

p2 = plot_ordination(ps2, GP.ord, type="TRAT2", color="TRAT2", shape="DEPTH") 
p2 + geom_polygon(aes(fill=TRAT2)) + geom_point(size=5) + ggtitle("samples")

### Loop through different method parameter options to the plot_ordination function and store the plot results in a list. This code chunk performs each ordination method, creates the corresponding graphic based on the first two axes of each ordination result, and then stores each ggplot2 plot object in a different named element of the list named plist:
dist = "bray"
ord_meths = c("CCA", "RDA", "NMDS", "MDS", "PCoA")
plist = llply(as.list(ord_meths), function(i, physeq, dist){
        ordi = ordinate(physeq, method=i, distance=dist)
        plot_ordination(physeq, ordi, "samples", color="TRAT2")
}, ps2, dist)

names(plist) <- ord_meths

### The following chunk will extract the data from each of those individual plots, and put it back together in one big data.frame suitable for including all plots in one graphic:
pdataframe = plyr::ldply(plist, function(x){
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe)[1] = "method"

p = ggplot(pdataframe, aes(Axis_1, Axis_2, color=TRAT2, shape=DEPTH, fill=TRAT2, alpha=.3))
p = p + geom_point(size=4) + geom_polygon()
p = p + facet_wrap(~method, scales="free")
p = p + scale_fill_manual(values=cols)
p = p + scale_colour_manual(values=cols)
p
#ggsave("ordination_methods.png", plot = p, width = 12, height = 10, dpi = 300, bg = "white")
```

*NMDS (Non-Metric Multidimensional Scaling)*
-A non-linear non-metric ordination technique.
-Ranks distances rather than preserving actual values.
-Only assumes that rank order of distances is meaningful.
-Axes are arbitrary and do not have intrinsic meaning.
-The distance between points is only relative (ordinal).
Best use case: When working with non-Euclidean distances (e.g., Bray-Curtis) and non-normal ecological data.
ðŸŸ¢ Advantages:
More robust for ecological data with non-linear patterns.
Works well with Bray-Curtis and other ecological dissimilarity metrics.
ðŸ”´ Disadvantages:
Cannot directly compare axes across different NMDS runs.
Stress values must be interpreted to determine fit (stress < 0.2 is acceptable, < 0.1 is ideal).

```{r NMDS}
nmds_data <- plist[[3]]$data  # NMDS is the third method in plist

# Rename axes for clarity
colnames(nmds_data)[1:2] <- c("Axis_1", "Axis_2")
nmds_data$DEPTH <- ordered(nmds_data$DEPTH, levels=c("topsoil", "subsoil"))

ggplot(nmds_data, aes(x = Axis_1, y = Axis_2, color = TRAT2, fill = TRAT2)) +
  geom_point(size = 4, alpha = 0.7) + 
  geom_polygon(alpha = 0.2) +
  scale_fill_manual(values = cols) + 
  scale_colour_manual(values = cols) +
  theme_minimal() +
  facet_wrap(~DEPTH)+
  labs(x = "NMDS Axis 1", y = "NMDS Axis 2")
#ggsave("nmds.png")

#################
OTU.ord <- ordinate(ps2, "NMDS", "bray")

ps2_meta_data <- data.frame(sample_data(ps2))  # Get the metadata from the ps object

plot_vars <- c("Slope_per","Organic_layer_mean_depth_cm", "Charred_organic_layer_mean_cm", "Green_per", "Brown_per", "Consumed_per", "Grass_per", "Old_litter_per", "New_litter_per", "Mineral_per", "Ash_per", "Stone_per")

env_df <- ps2_meta_data %>%
  dplyr::select(all_of(plot_vars))

envfit_results  <- vegan::envfit(OTU.ord, env_df, permutations = 999, na.rm = TRUE)

# Plot the ordination and add the environmental vectors
plot_ordination(ps2, OTU.ord, color = "TRAT2") +
  geom_point(size = 4) +  # Adjust point size
  geom_polygon(alpha = 0.2) +
  scale_fill_manual(values = cols) + 
  scale_colour_manual(values = cols) +
  geom_segment(data = as.data.frame(envfit_results$vectors$arrows), 
               aes(x = 0, y = 0, xend = NMDS1, yend = NMDS2),
               arrow = arrow(type = "open", length = unit(0.2, "inches")), 
               size = 1, color = "blue") +
  geom_text(data = as.data.frame(envfit_results$vectors$arrows), 
            aes(x = NMDS1, y = NMDS2, label = rownames(envfit_results$vectors$arrows)),
            size = 3, color = "blue") +
  theme_minimal()
#ggsave("nmds2.png")

#####TOPSOIL
ps2_topsoil <- subset_samples(ps2, DEPTH == "topsoil")

# Perform ordination for topsoil samples
OTU.ord_topsoil <- ordinate(ps2_topsoil, method = "NMDS", distance = "bray")

env_df_TOP <- ps2_meta_data %>%
  filter(DEPTH == "topsoil") %>%
  dplyr::select(all_of(plot_vars))

envfit_TOP  <- vegan::envfit(OTU.ord_topsoil, env_df_TOP, permutations = 999, na.rm = TRUE)

# Plot the ordination and add the environmental vectors
plot_ordination(ps2, OTU.ord_topsoil, color = "TRAT2") +
  geom_point(size = 4) +  # Adjust point size
  geom_polygon(alpha = 0.2) +
  scale_fill_manual(values = cols) + 
  scale_colour_manual(values = cols) +
  geom_segment(data = as.data.frame(envfit_TOP$vectors$arrows), 
               aes(x = 0, y = 0, xend = NMDS1, yend = NMDS2),
               arrow = arrow(type = "open", length = unit(0.2, "inches")), 
               size = 1, color = "blue") +
  geom_text(data = as.data.frame(envfit_TOP$vectors$arrows), 
            aes(x = NMDS1, y = NMDS2, label = rownames(envfit_TOP$vectors$arrows)),
            size = 3, color = "blue") +
  scale_x_reverse()+
  theme_minimal()
#ggsave("nmdsTOP_env.png")

#####SUBSOIL
ps2_subsoil <- subset_samples(ps2, DEPTH == "subsoil")

# Perform ordination for subsoil samples
OTU.ord_subsoil <- ordinate(ps2_subsoil, method = "NMDS", distance = "bray")

env_df_SUB <- ps2_meta_data %>%
  filter(DEPTH == "subsoil") %>%
  dplyr::select(all_of(plot_vars))

envfit_SUB  <- vegan::envfit(OTU.ord_subsoil, env_df_SUB, permutations = 999, na.rm = TRUE)

# Plot the ordination and add the environmental vectors
plot_ordination(ps2, OTU.ord_subsoil, color = "TRAT2") +
  geom_point(size = 4) +  # Adjust point size
  geom_polygon(alpha = 0.2) +
  scale_fill_manual(values = cols) + 
  scale_colour_manual(values = cols) +
  geom_segment(data = as.data.frame(envfit_SUB$vectors$arrows), 
               aes(x = 0, y = 0, xend = NMDS1, yend = NMDS2),
               arrow = arrow(type = "open", length = unit(0.2, "inches")), 
               size = 1, color = "blue") +
  geom_text(data = as.data.frame(envfit_SUB$vectors$arrows), 
            aes(x = NMDS1, y = NMDS2, label = rownames(envfit_SUB$vectors$arrows)),
            size = 3, color = "blue") +
  scale_x_reverse()+
  theme_minimal()
#ggsave("nmdsSUB_env.png")
```

```{r PERMANOVA}
permanova <- vegan::adonis2(phyloseq::distance(ps2, method="bray") ~ TRAT2*DEPTH,
             permutations = 999, data = ps2_meta_data)
permanova

####TOPSOIL
meta_data_top <- data.frame(sample_data(ps2_topsoil))
permanova_top <- vegan::adonis2(phyloseq::distance(ps2_topsoil, method="bray") ~ TRAT2,
                                permutations = 999, data = meta_data_top)
permanova_top

####SUBSOIL
meta_data_sub <- data.frame(sample_data(ps2_subsoil))
permanova_sub <- vegan::adonis2(phyloseq::distance(ps2_subsoil, method="bray") ~ TRAT2,
                                permutations = 999, data = meta_data_sub)
permanova_sub

#Model R2 = the proportion of total variation in the Bray-Curtis distances that is explained by TRAT2
#Residual R2: This is the proportion of variation not explained by TRAT2
```

```{r Indicator Species Analysis}
## Species are chosen as indicators if they:
# 1) Reflect the biotic or abiotic state of the environment;
# 2) Provide evidence for the impacts of environmental change; or
# 3) Predict the diversity of other species, taxa or communities within an area.

#Indicator species are often determined using an analysis of the relationship between the species occurrence or abundance values from a set of sampled sites and the classification of the same sites into site groups, which may represent habitat types, community types, disturbance states, etc. Thus, there are two data elements in an indicator species analysis: 
###############################
# (1) the community data matrix --> with sites in rows and species in columns
# (2) the vector that describes the classification of sites into groups --> a vector containing the classification of the sites into groups

df <-  as.data.frame((otu_table(ps2)))
head(df)

# Transpose the table
asv_transposed <- as.data.frame(t(df))
head(asv_transposed)
asv_transposed <- as.matrix(asv_transposed)

#Now we need to code the vector of site groups. We want to do 4 groups: UB, LS, HS1, HS2
head(ps2_meta_data$TRAT2)
groups4 <- as.factor(ps2_meta_data$TRAT2)

#Indicator species analysis
indval <- indicspecies::multipatt(asv_transposed, groups4, 
                    control = how(nperm=999)) # El argumento control = how(nperm=999) permite elegir el nÃºmero de permutaciones aleatorias requeridas para la prueba de permutaciÃ³n (este nÃºmero afecta la precisiÃ³n del valor p). La funciÃ³n how() del paquete permute permite definir diseÃ±os permutacionales mÃ¡s complejos
summary(indval) 

indval$sign
#The first columns of indicate (with ones and zeroes) which site groups were included in the combination preferred by the species. Then, the column indicates the index of the site group combination. The remaining two columns are the association statistic and the p-value of the permutational test.

# AÃ±adir indvalcomp = TRUE" a la funciÃ³n anterior nos brinda informaciÃ³n adicional sobre por quÃ© las especies pueden usarse como indicadores:
summary(indval, indvalcomp=TRUE)
# A value will tell us if the species is a good indicator of that group (1 = the best indicator)
# B value will tell us if the species appears in all the plots of the group 

# Al aÃ±adir alfa = 1, decimos que queremos mostrar el grupo al que estÃ¡ asociada cada especie, independientemente de si la asociaciÃ³n es significativa o no. (por defecto viene establecido alpha 0,05, que se corresponde al valor de significancia que  tiene en cuenta el anÃ¡lisis de permutaciÃ³n)  
summary(indval, indvalcomp=TRUE, alpha=1) 

# Extract significant species and their indicator values
significant_species <- indval$
indicator_values <- significant_species$sign

# Create a dataframe with species, groups, and indicator values
indicator_data <- data.frame(
  Species = rownames(indicator_values),
  Group = colnames(indicator_values),
  IndicatorValue = as.vector(indicator_values)
)



# Transform the row names into a column
asv_transposed <- rownames_to_column(asv_transposed, var = "Site")
# Convert all columns to numeric except the first one
asv_transposed <- (asv_transposed) %>%
  mutate(across(-Site, as.numeric))
head(asv_transposed)
# Paste some columns from the meta_data dataframe:
meta2 <- rownames_to_column(meta_data, var = "Site")
meta2 <- meta2 %>% 
  dplyr:: select(c(1:8))
asv_transposed_complete <- left_join(asv_transposed, meta2, by = "Site")
#write.csv(asv_transposed_complete, "asv_transposed_complete.csv")
asv_transposed_complete <- read.csv("F:/FPI - INIA/ADN/Firefungi_2021/asv_transposed_complete.csv", sep=";")



#######################################################################
#Now we want to do 8 groups: UB_topsoil, UB_subsoil, LS_topsoil, LS_subsoil, HS1_topsoil, HS1_subsoil, HS2_topsoil, HS2_subsoil
head(ps2_meta_data$TRAT2_DEPTH)
groups8 <- as.factor(ps2_meta_data$TRAT2_DEPTH)

#Indicator species analysis
indval8 <- indicspecies::multipatt(asv_transposed, groups8, 
                    control = how(nperm=999))
summary(indval8)
```

*PCoA (Principal Coordinate Analysis)*
- A linear metric ordination technique.
- Distance preservation: Attempts to preserve actual distances (Euclidean-like) as much as possible.
-Assumes that the dissimilarity metric is a good approximation of true distances.
-The axes are ordered based on the amount of variation they explain.
-The distance between points is quantitative and can be directly interpreted.
-Best use case: When you assume a linear relationship between distances and the underlying data structure.
ðŸŸ¢ Advantages:
Provides a direct visualization of group differences.
The percentage of variance explained by each axis is meaningful.
ðŸ”´ Disadvantages:
Can be influenced by highly skewed data or non-Euclidean relationships.
Assumes a Euclidean-like structure, which is not always valid for ecological data.

```{r Principal Coordinate Analysis (PCoA)}
bray_dist <- phyloseq::distance(ps2, method="bray")
# Perform PCoA
pcoa_result <- cmdscale(bray_dist, k = 2)  # k = dimensions
# Create a data frame for the PCoA results
pcoa_df <- data.frame(PCoA1 = pcoa_result[, 1],
                      PCoA2 = pcoa_result[, 2],
                      Sample = rownames(pcoa_result))
ps2_meta_data <- ps2_meta_data %>%
  mutate(Sample = rownames(ps2_meta_data))

# Merge with metadata
pcoa_df <- merge(pcoa_df, ps2_meta_data, by = "Sample")

# Plot PCoA
ggplot(pcoa_df, aes(x = PCoA1, y = PCoA2, color = TRAT2, shape = DEPTH)) +
  geom_point(size = 4) +
  scale_color_manual(values = cols) +  
  theme_minimal() +
  labs(x = "PCoA 1", y = "PCoA 2") +
  theme(legend.position = "none") 
#ggsave("PCoA.png")

##############################################################################
########### However our data does not seem Euclidean #########################
##############################################################################
#Zero-infltion check
sum(psmelt(ps2)$Abundance == 0) / nrow(psmelt(ps2)) #Our data is super zero-inflated

#Negative eigen values --> non-Euclidean data
bray_dist <- phyloseq::distance(ps2, method = "bray")
pcoa_bray <- ape::pcoa(bray_dist)
# Check for negative eigenvalues
pcoa_bray$values$Eigenvalues
# If any value is negative â†’ data is non-Euclidean

#############################################################################
#############################################################################

####TOPSOIL
bray_dist_top <- phyloseq::distance(ps2_topsoil, method="bray")
# Perform PCoA
pcoa_top <- cmdscale(bray_dist_top, k = 2)  # k = dimensions
# Create a data frame for the PCoA results
pcoa_top_df <- data.frame(PCoA1 = pcoa_top[, 1],
                      PCoA2 = pcoa_top[, 2],
                      Sample = rownames(pcoa_top))
meta_data_top <- meta_data_top %>%
  mutate(Sample = rownames(meta_data_top))

# Merge with metadata
pcoa_top_df <- merge(pcoa_top_df, meta_data_top, by = "Sample")

# Plot PCoA
ggplot(pcoa_top_df, aes(x = PCoA1, y = PCoA2, color = TRAT2)) +
  geom_point(size = 4) +
  scale_color_manual(values = cols) +  
  theme_minimal() +
  labs(x = "PCoA 1", y = "PCoA 2") +
  theme(legend.position = "none") 
#ggsave("PCoA_top.png")

####SUBSOIL
bray_dist_sub <- phyloseq::distance(ps2_subsoil, method="bray")
# Perform PCoA
pcoa_sub <- cmdscale(bray_dist_sub, k = 2)  # k = dimensions
# Create a data frame for the PCoA results
pcoa_sub_df <- data.frame(PCoA1 = pcoa_sub[, 1],
                      PCoA2 = pcoa_sub[, 2],
                      Sample = rownames(pcoa_sub))
meta_data_sub <- meta_data_sub %>%
  mutate(Sample = rownames(meta_data_sub))

# Merge with metadata
pcoa_sub_df <- merge(pcoa_sub_df, meta_data_sub, by = "Sample")

# Plot PCoA
ggplot(pcoa_sub_df, aes(x = PCoA1, y = PCoA2, color = TRAT2)) +
  geom_point(size = 4) +
  scale_color_manual(values = cols) +  
  theme_minimal() +
  labs(x = "PCoA 1", y = "PCoA 2") +
   scale_x_reverse()+
  theme(legend.position = "none") 
#ggsave("PCoA_sub.png")
```

```{r Hellinger transformation}
# It is designed to transform count data or relative abundance data in a way that reduces the influence of large values and makes the data more suitable for techniques like Principal Component Analysis (PCA) or cluster analysis. The transformation helps in emphasizing the relative differences between samples.
# It consist in 1) convert count data to proportions and 2) apply a square root transformation of 
# It preserves EUCLIDEAN DISTANCES --> perfect for Beta-diversity and other analysis

#Create a new phyloseq object:
ps_hel <- ps2
# Extract the OTU table
otu_table_ps <- otu_table(ps_hel)
taxa_rows <- taxa_are_rows(ps_hel)

# Perform Hellinger transformation
hellinger_otu_table <- vegan::decostand(as(otu_table_ps, "matrix"), method = "hellinger")
# Replace the OTU table in the phyloseq object
otu_table(ps_hel) <- otu_table(hellinger_otu_table, taxa_are_rows = taxa_rows)
```

```{r Principal Component Analysis (PCA)}

```






```{r Abundances}
##Total number of reads per ASV
total_abundance <- taxa_sums(ps1)
print(total_abundance)

abundance_df <- data.frame(
  ASV_OTU_ID = names(total_abundance),
  Total_Abundance = total_abundance)
head(abundance_df)
#write.csv(abundance_df, "abundance_df.csv")

## Relative abundance transformation. Transform counts to relative abundances:
ps3_rel <- transform_sample_counts(ps3, function(x){x/sum(x)}*100)
ps3meltrel <- psmelt(ps3_rel)

# Construct table
OTU_sample <-  as.data.frame((otu_table(sample_relabun_ASV)))
taxonomy_sample <- as.data.frame(tax_table(sample_relabun_ASV))
identical(rownames(OTU_sample),rownames(taxonomy_sample))
#> [1] TRUE
sample_relabun_ASV_table  <- cbind(taxonomy_sample, OTU_sample)
#SORT IT
sample_relabun_ASV_table <- sample_relabun_ASV_table[order(sample_relabun_ASV_table$ASV_names, decreasing = FALSE),]
#Check relative abundance sums 100
colSums(sample_relabun_ASV_table[,8:ncol(sample_relabun_ASV_table)])
##See table
head(sample_relabun_ASV_table)
```

```{r}
### GET TABLE BY SAMPLE AT GENUS LEVEL ###
# This chunk will give us information of the relative abundance of each ASV by treatment (eg.: we can see that ASV1 has the highest relative abundance for the HS2 while for the UB the most abundant is ASV10)

#Glom phyloseq
loc_phyloseq_genus <- tax_glom(ps3, taxrank = "Genus") #Change Genus to the desired taxonomic rank
sample_relabun_genus <- transform_sample_counts(loc_phyloseq_genus, function(x){x/sum(x)}*100)
##Extract elements from phyloseq
OTU_sample_genus <-  as.data.frame((otu_table(sample_relabun_genus)))
taxonomy_sample_genus  <- as.data.frame(tax_table(sample_relabun_genus)[,-7])
identical(rownames(OTU_sample_genus),rownames(taxonomy_sample_genus))
#> [1] TRUE
#Bind table
sample_table_genus  <- cbind(taxonomy_sample_genus, OTU_sample_genus)
#Sort table from most to least abundant genera (calculated as sum of abundance between samples)
sample_table_genus_final <- sample_table_genus [order(rowSums(sample_table_genus [,7:ncol(sample_table_genus )]), decreasing=TRUE),]
#See first lines of this table
head(sample_table_genus_final)

### GROUP GENUS TABLE BY LOCATION/TREATMENT
#Save genus abundances
otu_genus <- sample_table_genus_final[,7:ncol(sample_table_genus_final)] %>% t() %>% as.data.frame()
#Save taxonomy data
tax_genus <- sample_table_genus_final[,1:6]
#Calculate OTU mean abundance based on grouping factor (e.g., location)
location_mean_genus <- aggregate(otu_genus, by=list(meta_data$TRAT2), FUN=mean)%>% column_to_rownames("Group.1") %>% t() #Change meta_data and TRAT2 to your metadata and variable of interest
#Calculate OTU SD  based on grouping factor (e.g., location) and change colnames
location_SD_genus <- aggregate(otu_genus, by=list(meta_data$TRAT2), FUN=sd)%>% column_to_rownames("Group.1")  %>% t()  %>% #Change metadata.micro and location to your metadata and variable of interest
as.data.frame() %>% rename_with(.fn= ~paste0(colnames(location_mean_genus), "SD"))
#Merge mean abundance, SD and taxonomy.
genus_location_table <- merge(tax_genus, location_mean_genus, by=0) %>%column_to_rownames("Row.names") %>%
merge(location_SD_genus, by=0) %>% column_to_rownames("Row.names")
#Check abundances sum 100
colSums(genus_location_table[,7:ncol(genus_location_table)])
#Sort table from most to least abundant genera (calculated as sum of abundance between samples)
genus_location_table <- genus_location_table[order(rowSums(genus_location_table[,7:ncol(genus_location_table)]), decreasing=TRUE),]
#View table
head(genus_location_table)
```

```{r Abundances Violin: all phyla}
plot_abundance = function(physeq, ylabn = "", 
                          Facet = "Phylum", 
                          Color = "Phylum"){ 
  mphyseq = psmelt(physeq) 
  mphyseq <- subset(mphyseq, Abundance > 0) 
  ggplot(data = mphyseq, 
         mapping = aes_string(x = "TRAT2", y = "Abundance", 
                              color = Color, fill = Color)) + 
    geom_violin(fill = NA) + 
    geom_point(size = 1, alpha = 0.3, position = position_jitter(width = 0.3)) + 
    facet_wrap(facets = Facet) + 
    ylab(ylabn) + 
    scale_y_log10() }

plot_abundance(ps3,"Original Abundances") 
#ggsave("abun_ps3.png")
```

```{r Abundances: Top9 Genus}
# We need just the most abundant taxa, in this case the top 9
top9 <-microbiome::top_taxa(ps3, n = 9) # top_taxa returns n most abundant taxa, sorted by abundance
top9
ps_5 <- prune_taxa(top9, ps3)

ps5melted <- psmelt(ps_5)

cols_3 <- c("#41b6c4","#fd8d3c","#cb181d","#800026")
ps5melted$DEPTH <- factor(ps5melted$DEPTH, levels = c("topsoil", "subsoil"))
ggplot(ps5melted, aes(x= TRAT2, y=Genus, size = Abundance, color=TRAT2)) +
  geom_point()+
  scale_size(range = c(.1, 15)) +
  scale_color_manual(values = cols_3) +
theme(
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black"),
    axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    strip.placement = "outside",  # Place strip labels outside the plot
    strip.background = element_blank(),
    legend.position = "none" # Remove strip background
  ) +
  facet_wrap(~DEPTH)
#ggsave("Genus_bubbles2.png")
```

```{r Sankey Diagram}
## We need two dataframes:
# Nodes: with a number for each node we want to represent (ins and outs) --> e.g.: 1-UB, 2-LS, 3-HS1... Two columns
# Links: Three columns. Source (from 0 to nÂº nodes - 1), Target and Value
h <- ps3meltrel %>%
  dplyr::group_by(TRAT2, DEPTH, Phylum) %>%
  dplyr::summarise(total_abundance = sum(Abundance, na.rm = TRUE))

#write.csv(h, "ps4melted.csv") --> we make the codification in Excel

links <- as.data.frame(read_excel("C:/Users/amara.santiesteban/OneDrive - UAM/Escritorio/Sankey_trial.xlsx", sheet = "Links"))
nodes <- as.data.frame(read_excel("C:/Users/amara.santiesteban/OneDrive - UAM/Escritorio/Sankey_trial.xlsx", sheet = "Nodes"))

sankey <- networkD3::sankeyNetwork(Links = links, Nodes = nodes, Source = "source",
               Target = "target", Value = "Value", NodeID = "names",  iterations = 0) #Setting "iterations = 0" will disable the algorithm which automatically determines the node placement, and the nodes will be placed in the order that they appear in the Nodes dataframe.

pdf_file <- "sankey_diagram.pdf"
webshot2::webshot(html_file, file = pdf_file)
```

```{r Comparation with BLAST}
#top5_UNITE22 <- as.data.frame(read_xlsx("Libro1.xlsx", col_names=TRUE, sheet = 1))
head(top5_UNITE22)

ggplot(top5_UNITE22, aes(fill=Family2, y=Count, x=TRAT2)) + 
    geom_bar(position="stack", stat="identity")

top5_UNITE23 <- as.data.frame(read_xlsx("Libro1.xlsx", col_names=TRUE, sheet = 2))
```

```{r Rarefaction curves}
devtools::install_github("gauravsk/ranacapa")

plot <- ranacapa::ggrare(ps1, step = 100, label = NULL, color = "ID", se = TRUE)
#(plot <- plot + facet_wrap(~TRAT2))
#ggsave("rarefactionps3.png")
```