---
title: "Phyloseq analysis of Navalacruz 2021 dataset sampling"
author: "Amara Santiesteban Serrano"
date: " r format(Sys.time(), '%d %B, %Y') "
output: 
  html_document:
    df_print: paged
    pdf_document: default
    word_document: default
  editor_options:
    chunk_output_type: inline
---

*Primero procesar las secuencias en DADA2 (1Âº) y en LULU (2Âº)*

```{r Packages}
#install.packages("pacman")
pacman::p_load("ggplot2","tidyr","dplyr","tidyverse","readxl","pastecs","lattice","forcats","corrgram","corrplot","HH","effects","car","multcompView","lme4", "devtools","emmeans","factoextra","ggfortify","RColorBrewer","MuMIn","scales","multifunc", "ggthemes", "knitr", "BiocManager", "phyloseq", "phangorn", "DECIPHER", "biomformat", "rbiom", "plyr", "indicspecies", "networkD3", "htmlwidgets", "webshot2", "vegan", "lme4", "car", "plyr", "ggpubr", "ggsignif", "ape", "cooccur", "igraph", "reshape2", "cowplot", "patchwork")
```

```{r Packages: BiocManager}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("microbiome")
BiocManager::install(c("DECIPHER", "phangorn"))

remotes::install_github("gavinsimpson/ggvegan")
```

```{r Colors}
cols <- c("#41b6c4","#fd8d3c","#cb181d","#800026")
```

```{r Phylogenetic Tree}
# We will generated from the final sequences table (from DADA2)
asv_seqs <- colnames(seqtab.nochim)
asv_names <- paste0("ASV", seq_along(asv_seqs))

# Create DNAStringSet
library(Biostrings)
asv_dna <- DNAStringSet(asv_seqs)
names(asv_dna) <- asv_names

#To work with the *phangorn* package to build a tree the first step is reading an alignment:
# Align sequences
alignment <- AlignSeqs(asv_dna, anchor = NA)

#After reading in the nucleotide alignment we can build a first tree with distance based methods. 
# Create a distance matrix
phang_align <- phyDat(as(alignment, "matrix"), type = "DNA")
dm <- dist.ml(phang_align)

# Build tree
treeUPGMA  <- upgma(dm) # Is a simple agglomerative (bottom-up) hierarchical clustering method --> rooted
#plot(treeUPGMA, main="UPGMA")
treeNJ <- NJ(dm) # Neighbor-Joining tree --> unrooted
#plot(treeNJ, "unrooted", main="NJ")

# Bootstrap
fun <- function(x) upgma(dist.ml(x))
bs_upgma <- bootstrap.phyDat(phang_align,  fun)


```

```{r data required}
#########################
#OTU abundance: *matrix*            class(clustered_asv_table)
#Taxonomy: *matrix*                 class(clustered_taxa)
#Metadata: *data.frame*
#FunGuild: *matrix*
########################

#We need the taxa table and the curated asv table to have the same number of entries
#We need the meta_data and the curated asv table to have the exact same samples names

#####OTU abundance:
colnames(clustered_asv_table) <- sapply(colnames(clustered_asv_table), function(name) {
  parts <- strsplit(name, "-FIREFUNGI")[[1]]
  return(paste0(parts[1], "-FIREFUNGI"))
})
#clustered_asv_table[] <- lapply(clustered_asv_table, as.numeric)

####Taxonomy: 
class(clustered_tax_table)
clustered_tax_table <- as.matrix(clustered_tax_table)

####Metadata
meta_data <- as.data.frame(Firefungi_soil_dataset_2021)
#We want to add a gradient of severity (pc1)
plot_vars <- c("Slope_per","Organic_layer_mean_depth_cm", "Charred_organic_layer_mean_cm", "Green_per", "Brown_per", "Consumed_per", "Grass_per", "Old_litter_per", "New_litter_per", "Mineral_per", "Ash_per", "Stone_per")
pca_sev <- prcomp(meta_data[,plot_vars], scale=T)
pc1 <- pca_sev$x[,1] 
pc1 <- as.data.frame(pc1)
meta_data_pc1 <- meta_data %>%
  dplyr::mutate(pc1 = pc1$pc1)
rownames(meta_data_pc1) <- paste0(1:80, "-FIREFUNGI")
meta_data_pc1 <- meta_data_pc1 %>% 
  mutate(Fe_g_Kg = Fe_mg_Kg_105/1000,
         Ergosterol_Âµg_g = Ergosterol_ng_g_105/1000,
         C_N_ratio = Ctotal_g_100g_105/Ntotal_g_100g_105)

####Phylogenetic Tree

```

The *phyloseq package* is a tool to import, store, analyze, and graphically display complex phylogenetic sequencing data that has already been clustered into Operational Taxonomic Units (OTUs), especially when there is associated sample data, phylogenetic tree, and/or taxonomic assignment of the OTUs. This package leverages many of the tools available in R for ecology and phylogenetic analysis (vegan, ade4, ape, picante), while also using advanced/flexible graphic systems (ggplot2) to easily produce publication-quality graphics of complex phylogenetic data. 

```{r Phyloseq object}
ps <- phyloseq::phyloseq(otu_table(clustered_asv_table, taxa_are_rows=TRUE), 
               sample_data(meta_data, sample_names(sample.names)),
               tax_table(clustered_taxa))
ps
```

```{r Quality analysis}
# Define prevalence of each taxa (in how many samples did each taxa appear at least once)
# Compute prevalence (number of samples where an OTU is present)
prevdf <- apply(X = otu_table(ps),
                MARGIN = ifelse(taxa_are_rows(ps), yes = 1, no = 2),
                FUN = function(x) sum(x > 0))

prevdf <- data.frame(Prevalence = prevdf, row.names = taxa_names(ps))

# Add total abundance and taxonomy information
prevdf <- data.frame(Prevalence = prevdf$Prevalence,
                     TotalAbundance = phyloseq::taxa_sums(ps),
                     as.data.frame(tax_table(ps)))

#With the following table we can see how well represented is each "Phylum" (or Kingdom, etc...)
dfprev <- plyr::ddply(prevdf, "Kingdom", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
```

```{r Visualization}
# Define prevalence threshold as 5% of total samples 
prevalenceThreshold = 0.05 * phyloseq::nsamples(ps1)
prevalenceThreshold

#This plot will help us to identify which taxa are more abundant and prevalent in the dataset.
ggplot(prevdf, aes(TotalAbundance, Prevalence, color = Phylum)) + 
  geom_hline(yintercept = prevalenceThreshold, alpha = 0.5, linetype = 2) + 
  geom_point(size = 2, alpha = 0.7) + 
  scale_y_log10() + 
  scale_x_log10() + 
  xlab("Total Abundance") + 
  facet_wrap(~Phylum) +
  theme(legend.position = "none")
#ggsave("prevalence.png")

#This will show us the distribution of read counts per sample:
sample_sum_df <- data.frame(sum = sample_sums(ps1))
ggplot(sample_sum_df, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "grey", binwidth = 2500) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  theme(axis.title.y = element_blank()) 
```

```{r ps0}
# We make sure that we do not have OTUs that are not present in any of the samples:
ps0 <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
ps0
#We also did a filtering of ASV with less than 5 reads but the result is the same (probably due to LULU & postclustering)
```

```{r Filtering: unwanted taxa}
#Based on the results above, we could decide to filter some Phylum out (the ones with less representation). We can also get rid of the Kingdoms not corresponding to fungi:
#plot_bar(ps, fill = "Kingdom") + 
#geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack")

ps1 = phyloseq::subset_taxa(ps, Kingdom=="k__Fungi")
ps1

#plot_bar(ps1, fill = "Kingdom") + 
#geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack")
```

*ADVICE*: check some sequences (from the 'asv_OTU.fasta' generated in DADA2, they can be accessed through Notepad++) in BLAST: https://blast.ncbi.nlm.nih.gov/Blast.cgi# --> copy and paste the whole sequence in the Entry query sequence box, and click on 'BLAST' at the end. This is a double check to see if our reference database (UNITE in this case) has assigned correctly the taxons.

```{r Alpha Diversity}
# Observed = Richness: total number of OTU observed in a sample (it does not consider abundance or eveness)
# Chao1: non-parametric estimator of the minimum species richness. The number of species in a community, accounting for the number of rare species that are likely to be missed
# Shannon: It measures the diversity of a community by considering both species richness and evenness. Higher values indicate greater diversity, with more even distribution of species abundances.

alpha_fungi <- microbiome::alpha(ps1, index = c("Observed", "Shannon", "Chao1")) #ps1: only for Fungi Kingdom

alpha_fungi <- alpha_fungi %>% 
  tibble::rownames_to_column(var = "SampleID")

metadata_df <- meta_data_pc1 %>% 
  tibble::rownames_to_column(var = "SampleID")

alpha_fungi <- left_join(alpha_fungi, metadata_df, by = "SampleID")
```

```{r Alpha Diversity Models}
alpha_fungi$TRAT2 <- ordered(alpha_fungi$TRAT2, levels=c("UB", "LS", "HS1", "HS2"))
alpha_fungi$DEPTH <- ordered(alpha_fungi$DEPTH, levels=c("topsoil", "subsoil"))

### Visually assessment of data
var <- alpha_fungi$chao1

hist(var) 

alpha_fungi %>%
  ggplot(aes(x=TRAT2, y=var, fill=DEPTH)) +
  geom_boxplot(outlier.color = "red",
    outlier.shape = 8) +
  scale_fill_brewer(palette="Dark2") +
  stat_summary(fun= mean, geom="point", shape=20, size=2, colour="black") + 
  facet_wrap(~ DEPTH, scales="free_x") 

#################### WHOLE ##################################
#### FIRST!! We define the Contrast table we want to work with
options(contrasts=c(factor="contr.sum", ordered="contr.poly"))
##############################################################
mixed.null_SITE <- lmer(var ~ (1|SITE), data = alpha_fungi)
mixed.null_PLOT <- lmer(var ~ (1|PLOT), data = alpha_fungi)
mixed.null_COMBI <- lmer(var ~ (1|SITE*PLOT), data = alpha_fungi)
##############################################################
#We want to make sure which one should be the random factor
AIC(mixed.null_SITE, mixed.null_PLOT, mixed.null_COMBI)
##############################################################

mixed.var <- lmer(var ~ TRAT2*DEPTH + (1|SITE), data = alpha_fungi)  #SITE selected as random effect

par(mfrow=c(1,3))
qqnorm(resid(mixed.var)); qqline(resid(mixed.var))
hist(resid(mixed.var))
plot(fitted(mixed.var), resid(mixed.var)); abline(h=0)
par(mfrow=c(1,1))

car::Anova(mixed.var, test.statistic="F", type=3)

multcomp::cld(emmeans(mixed.var, ~TRAT2, type="response"), Letters= letters) #Tuckey's post-hoc test
```

*We've run Alpha Diversity with all ASV (ps) and with only Kingdom Fungi (ps1)*

-No significant differences found across *DEPTHS*
-No significant differences in *Shannon diversity* across treatments
-Significant differences found in *Observed and Chao1 richness*: 
  --> ps Observed and Chao1 richness (no sig. differences across DEPTHs; across treatmentsUB, UB > LS > HS1 > HS2)
  --> ps1 (k__Fungi): Observed and Chao1 richness (no sig. differences across DEPTHs; across treatments UB, LS > HS1, HS2)

```{r Alpha Diversity Plot}
# Reorder the levels of TRAT2 and DEPTH
sample_data(ps1)$TRAT2 <- factor(sample_data(ps1)$TRAT2, levels = c("UB", "LS", "HS1", "HS2") )
sample_data(ps1)$DEPTH <- factor(sample_data(ps1)$DEPTH, levels = c("topsoil","subsoil"))

plot_richness(ps1, color = "TRAT2", shape ="DEPTH", x = "TRAT2", 
              measures = c("Observed", "Chao1","Shannon")) + 
  geom_boxplot(aes(fill = TRAT2), alpha=.2, outlier.shape = NA) + 
  geom_jitter(width = 0.2, size = 1.5) +  # <- los puntos individuales
  scale_color_manual(values = c("#1d91c0","#feb24c","#cb181d","#800026")) + 
  scale_fill_manual(values = c("#1d91c0","#feb24c","#cb181d","#800026")) +
  #facet_grid(variable ~ DEPTH, scales="free") +
  theme(
    panel.grid = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid.minor = element_blank(),
    axis.line = element_line(),
    strip.background = element_blank(),  # Remove strip background
    axis.text.x = element_text(angle = 0.1, hjust = 1),  # Rotate x-axis labels
    panel.background = element_blank(),  # Remove panel background color
    plot.background = element_blank(),   # Remove plot background color
    #legend.position = "none" # Remove legend
  )
#ggsave("alpha_div_k_fungi.png")

ggplot(alpha_fungi, aes(x=TRAT2, y = observed, alpha=.6, color=TRAT2))+
  geom_point()+
  geom_boxplot(aes(fill = TRAT2)) + 
  scale_color_manual(values = c("#1d91c0","#feb24c","#cb181d","#800026")) + 
  scale_fill_manual(values = c("#1d91c0","#feb24c","#cb181d","#800026")) +
  theme_bw()+
  theme(
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(),
    strip.background = element_blank(),  # Remove strip background
    axis.text.x = element_text(angle = 0.1, hjust = 1),  # Rotate x-axis labels
    panel.background = element_blank(),  # Remove panel background color
    plot.background = element_blank(),   # Remove plot background color
    legend.position = "none"             # Remove legend
  )+
  facet_grid(~DEPTH)
#ggsave("observed_div_k_fungi.png")
```

```{r Agglomerate by taxonomic ranks}
# This step is necessary when we work with taxonomic ranks above ASV
# Phylum -> For example, in this case we need 2 numbers per sample (UB_TOPSOIL_1: one number for Ascomycota and one for Basidio)
ps1_phylum <- tax_glom(ps1, taxrank = "Phylum")
sample_data(ps1_phylum)
tax_table(ps1_phylum)
ps1_phylum_melted <- psmelt(ps1_phylum)

# Class
ps1_class <- tax_glom(ps1, taxrank = "Class")
sample_data(ps1_class)
tax_table(ps1_class)
ps1_class_melted <- psmelt(ps1_class)

# Order
ps1_order <- tax_glom(ps1, taxrank = "Order")
sample_data(ps1_order)
tax_table(ps1_order)
ps1_order_melted <- psmelt(ps1_order)

# Family
ps1_family <- tax_glom(ps1, taxrank = "Family")
sample_data(ps1_family)
tax_table(ps1_family)
ps1_family_melted <- psmelt(ps1_family)

# Genus
ps1_genus <- tax_glom(ps1, taxrank = "Genus")
sample_data(ps1_genus)
tax_table(ps1_genus)
ps1_genus_melted <- psmelt(ps1_genus)
```

```{r Ascomycota/Basidiomycota}
# Filter to include only specified phyla
filterPhylaAB <- c("p__Ascomycota", "p__Basidiomycota")
ps_ascobasidio = phyloseq::subset_taxa(ps1_phylum, Phylum %in% filterPhylaAB) #New Phyloseq object
sample_data(ps_ascobasidio)

#plot_bar(ps_ascobasidio, "Phylum", fill = "Phylum", facet_grid = ~TRAT2)

ascobasidio_melted <- psmelt(ps_ascobasidio)

####### Summarize the data to calculate mean and SE 
summary_data_ab <- ascobasidio_melted %>%
  dplyr::group_by(Phylum, TRAT2, DEPTH) %>%
  dplyr::summarise(
    mean_abundance = mean(Abundance, na.rm = TRUE),
    true_mean = sum(Abundance)/n(),
    count = n(),
    sum_abundance = sum(Abundance),
    sd_abundance = sd(Abundance),
    se_abundance = sd(Abundance) / sqrt(n()))  # Calculate standard error

# Set factors for DEPTH and TRAT2
summary_data_ab$DEPTH <- factor(summary_data_ab$DEPTH, levels = c("topsoil", "subsoil"))
summary_data_ab$TRAT2 <- factor(summary_data_ab$TRAT2, levels = c("UB", "LS", "HS1", "HS2"))
summary_data_ab$Phylum <- factor(summary_data_ab$Phylum, levels = c("p__Basidiomycota", "p__Ascomycota"))

cols_2 <- c("#fed976","#fd8d3c")

# Create the plot
ggplot(summary_data_ab, aes(x = Phylum, y = mean_abundance)) +
  geom_bar(aes(fill = Phylum), stat = "identity", position = "dodge") + 
  geom_errorbar(aes(ymin = mean_abundance - se_abundance, ymax = mean_abundance + se_abundance), width = 0.2, position = position_dodge(0.9)) +
  scale_fill_manual(values = cols_2) +
  facet_grid(DEPTH ~ TRAT2) +
  theme_bw()+
  theme(
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black"),
    axis.text.x = element_blank(),
    axis.ticks = element_line(color = "black"),
    strip.text = element_text(size = 12),
    strip.background = element_blank())  # Remove strip background
    #legend.position = "none"
#ggsave("AscoBasidio_meanabundance.png")
```

```{r Ascomycota/Basidiomycota Models}
# Fit a Linear Mixed-Effects Model
# Here, random effect is Sample (or any grouping variable) and fixed effects are Phylum, DEPTH, and TRAT2

lmm_results <- lmer(Abundance ~ Phylum * DEPTH * TRAT2 + (1|Sample), data = ascobasidio_melted)

par(mfrow=c(1,3))
qqnorm(resid(lmm_results)); qqline(resid(lmm_results))
hist(resid(lmm_results))
plot(fitted(lmm_results), resid(lmm_results)); abline(h=0)
par(mfrow=c(1,1))

car::Anova(lmm_results, test.statistic="F", type=3)

multcomp::cld(emmeans(lmm_results, ~TRAT2*DEPTH, type="response"), Letters=letters) #Tuckey's post-hoc test

#NO SIGNIFICANT DIFFERENCES AMONG FIRE SEVERITIES OR DEPTHS
```

```{r Most abundant 6 Order}
order.sum = tapply(phyloseq::taxa_sums(ps1_order), 
                   tax_table(ps1_order)[, "Order"], 
                   sum, na.rm=TRUE)
top6order = names(sort(order.sum, TRUE))[1:6]
psTOP_order = phyloseq::prune_taxa((tax_table(ps1_order)[, "Order"] %in% top6order), ps1_order)
psTOP_order

plot_bar(psTOP_order, "Order", fill = "Order", facet_grid = ~TRAT2) 

df_bar <- psmelt(psTOP_order)
summary_df <- df_bar %>%
    dplyr::group_by(Order, TRAT2, DEPTH) %>%
    dplyr::summarise(
    mean_abundance = mean(Abundance),
    sum_abundance = sum(Abundance),
    se_abundance = sd(Abundance) / sqrt(n()))

# fix order of factors
summary_df$TRAT2 <- factor(summary_df$TRAT2, levels = c("UB", "LS", "HS1", "HS2"))
summary_df$DEPTH <- factor(summary_df$DEPTH, levels = c("topsoil", "subsoil"))

# Color palette (10 orders)
cols6 <- RColorBrewer::brewer.pal(6, "Paired")

# Barplot
ggplot(summary_df, aes(x = Order, y = mean_abundance, fill = Order)) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_errorbar(aes(ymin = mean_abundance - se_abundance,
                    ymax = mean_abundance + se_abundance),
                    width = 0.2, position = position_dodge(0.9)) +
  facet_grid(DEPTH ~ TRAT2) +
  scale_fill_manual(values = cols6) +
  theme_bw() +
  theme(
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.ticks.x = element_blank(),
    strip.text = element_text(size = 12),
    #legend.position = "none"
  ) 
#ggsave("top6orders.png")
```

```{r Species co-ocurrence: Order}
tax_mat <- as.data.frame(tax_table(ps1_order))
otu_mat <- as.data.frame(otu_table(ps1_order)) 

# Create a named vector: names = ASV IDs, values = Orders
asv_to_order <- tax_mat$Order
names(asv_to_order) <- rownames(tax_mat)

# Convert abundance data to presence/absence (1 if >0, else 0)
asv_pa <- ifelse(otu_mat > 0, 1, 0)
asv_pa_order <- asv_pa
rownames(asv_pa_order) <- asv_to_order[rownames(asv_pa)]  # rename rows

cooccur_result <- cooccur(asv_pa_order, type = "spp_site",
                          thresh = TRUE, spp_names = TRUE)

#png("cooccurrence_plot.png", width = 800, height = 800) 
par(mfrow=c(2,1))
plot(cooccur_result)
#dev.off()

plot.cooccur(cooccur_result)
```

```{r Most abundant 10 Genus}
order.sum = tapply(phyloseq::taxa_sums(ps1_genus), 
                   tax_table(ps1_genus)[, "Genus"], 
                   sum, na.rm=TRUE)
top10genus = names(sort(order.sum, TRUE))[1:10]
psTOP_genus = phyloseq::prune_taxa((tax_table(ps1_genus)[, "Genus"] %in% top10genus), ps1_genus)
psTOP_genus

plot_bar(psTOP_genus, "Genus", fill = "Genus", facet_grid = ~TRAT2) 

df_bar <- psmelt(psTOP_genus)
summary_df <- df_bar %>%
    dplyr::group_by(Genus, TRAT2, DEPTH) %>%
    dplyr::summarise(
    mean_abundance = mean(Abundance),
    sum_abundance = sum(Abundance),
    se_abundance = sd(Abundance) / sqrt(n()))

# fix order of factors
summary_df$TRAT2 <- factor(summary_df$TRAT2, levels = c("UB", "LS", "HS1", "HS2"))
summary_df$DEPTH <- factor(summary_df$DEPTH, levels = c("topsoil", "subsoil"))

# Color palette (10 orders)
cols10 <- RColorBrewer::brewer.pal(10, "Paired")

# Barplot
ggplot(summary_df, aes(x = Genus, y = mean_abundance, fill = Genus)) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_errorbar(aes(ymin = mean_abundance - se_abundance,
                    ymax = mean_abundance + se_abundance),
                    width = 0.2, position = position_dodge(0.9)) +
  facet_grid(DEPTH ~ TRAT2) +
  scale_fill_manual(values = cols10) +
  theme_bw() +
  theme(
    #panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.ticks.x = element_blank(),
    strip.text = element_text(size = 12),
    #legend.position = "none"
  ) 
#ggsave("top10genus.png")
```

Plot.cooccur function available at: https://github.com/griffithdan/cooccur/blob/master/R/plot.cooccur.R

```{r Species co-ocurrence: Genus}
tax_mat <- as.data.frame(tax_table(ps1_genus))
otu_mat <- as.data.frame(otu_table(ps1_genus)) 

# Create a named vector: names = ASV IDs, values = Orders
asv_to_order <- tax_mat$Genus
names(asv_to_order) <- rownames(tax_mat)

# Convert abundance data to presence/absence (1 if >0, else 0)
asv_pa <- ifelse(otu_mat > 0, 1, 0)
asv_pa_order <- asv_pa
rownames(asv_pa_order) <- asv_to_order[rownames(asv_pa)]  # rename rows

cooccur_result <- cooccur(asv_pa_order, type = "spp_site",
                          thresh = TRUE, spp_names = TRUE)

#png("cooccurrence_plot.png", width = 800, height = 800) 
par(mfrow=c(2,1))
plot(cooccur_result)
#dev.off()
```

```{r Buubles top 10 Order}
ggplot(df_bar, aes(x= TRAT2, y=Order, size = Abundance, color=TRAT2)) +
  geom_point()+
  scale_size(range = c(.1, 15)) +
  scale_color_manual(values = cols) +
theme(
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black"),
    axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    strip.placement = "outside",  # Place strip labels outside the plot
    strip.background = element_blank(),
    legend.position = "none" # Remove strip background
  ) +
  facet_wrap(~DEPTH)
#ggsave("Bubbles_top10phyla.png")
```

*Beta diversity* is going to account for the *turnover* across sites, treatments...
It is based on distances --> we work with the *Hellinger transformed data*
*Bray-Curtis* dissimilarity is indicated for ABUNDANCE data 

```{r Hellinger transformation}
# It is designed to transform count data or relative abundance data in a way that reduces the influence of large values and makes the data more suitable for techniques like Principal Component Analysis (PCA) or cluster analysis. The transformation helps in emphasizing the relative differences between samples.
# It consist in 1) convert count data to proportions and 2) apply a square root transformation of 
# It preserves EUCLIDEAN DISTANCES --> perfect for Beta-diversity and other analysis

#Create a new phyloseq object:
ps_hel <- ps1
# Extract the OTU table
otu_table_ps <- otu_table(ps_hel)
taxa_rows <- taxa_are_rows(ps_hel)

# Perform Hellinger transformation
hellinger_otu_table <- vegan::decostand(as(otu_table_ps, "matrix"), method = "hellinger")
# Replace the OTU table in the phyloseq object
otu_table(ps_hel) <- otu_table(hellinger_otu_table, taxa_are_rows = taxa_rows)
```

```{r Beta diversity}
dist_hel <- vegdist(hellinger_otu_table, method="bray")
#heatmap(as.matrix(dist_hel))

#Plot of the different ordination methods
GP.ord <- ordinate(ps_hel, "NMDS", "bray")
p1 = plot_ordination(ps_hel, GP.ord, type="taxa", color="Phylum", title="taxa")
print(p1)

p1 + facet_wrap(~Phylum, 3)

p2 = plot_ordination(ps_hel, GP.ord, type="TRAT2", color="TRAT2", shape="DEPTH") 
p2 + geom_polygon(aes(fill=TRAT2)) + geom_point(size=5) + ggtitle("samples")

### Loop through different method parameter options to the plot_ordination function and store the plot results in a list. This code chunk performs each ordination method, creates the corresponding graphic based on the first two axes of each ordination result, and then stores each ggplot2 plot object in a different named element of the list named plist:
dist = "bray"
ord_meths = c("CCA", "RDA", "NMDS", "MDS", "PCoA")
plist = llply(as.list(ord_meths), function(i, physeq, dist){
        ordi = ordinate(physeq, method=i, distance=dist)
        plot_ordination(physeq, ordi, "samples", color="TRAT2")
}, ps_hel, dist)

names(plist) <- ord_meths

### The following chunk will extract the data from each of those individual plots, and put it back together in one big data.frame suitable for including all plots in one graphic:
pdataframe = plyr::ldply(plist, function(x){
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe)[1] = "method"

p = ggplot(pdataframe, aes(Axis_1, Axis_2, color=TRAT2, shape=DEPTH, fill=TRAT2, alpha=.3))
p = p + geom_point(size=4) + geom_polygon()
p = p + facet_wrap(~method, scales="free")
p = p + scale_fill_manual(values=cols)
p = p + scale_colour_manual(values=cols)
p
#ggsave("ordination_methods_ps1.png", plot = p, width = 12, height = 10, dpi = 300, bg = "white")
```

*PCoA (Principal Coordinate Analysis)*
- A linear metric ordination technique.
- Distance preservation: Attempts to preserve actual distances (Euclidean-like) as much as possible.
-Assumes that the dissimilarity metric is a good approximation of true distances.
-The axes are ordered based on the amount of variation they explain.
-The distance between points is quantitative and can be directly interpreted.
-Best use case: When you assume a linear relationship between distances and the underlying data structure.
ðŸŸ¢ Advantages:
Provides a direct visualization of group differences.
The percentage of variance explained by each axis is meaningful.
ðŸ”´ Disadvantages:
Can be influenced by highly skewed data or non-Euclidean relationships.
Assumes a Euclidean-like structure, which is not always valid for ecological data.

```{r Principal Coordinate Analysis (PCoA)}
bray_dist <- phyloseq::distance(ps_hel, method="bray")
# Perform PCoA
pcoa_result <- cmdscale(bray_dist, k = 2)  # k = dimensions
# Create a data frame for the PCoA results
pcoa_df <- data.frame(PCoA1 = pcoa_result[, 1],
                      PCoA2 = pcoa_result[, 2],
                      Sample = rownames(pcoa_result))
ps_hel_meta_data <- ps_hel_meta_data %>%
  mutate(Sample = rownames(ps_hel_meta_data))

# Merge with metadata
pcoa_df <- merge(pcoa_df, ps_hel_meta_data, by = "Sample")

#Add envfit
env2_df <- ps_hel_meta_data %>%
  dplyr::select(all_of(chem_vars))

envfit_results  <- vegan::envfit(pcoa_result, env2_df, permutations = 999, na.rm = TRUE)

arrows_df <- as.data.frame(scores(envfit_results, display = "vectors"))
arrows_df$Variable <- rownames(arrows_df)
mult <- 1.5  # para ajustar la longitud de las flechas al grÃ¡fico
arrows_df <- arrows_df %>%
  dplyr::mutate(PCoA1 = Dim1 * mult,
                PCoA2 = Dim2 * mult)

# Plot PCoA
pcoa_df %>%
  ggplot(aes(x = PCoA1, y = PCoA2, color = TRAT2, shape = DEPTH)) +
  geom_point(size = 4) +
  geom_segment(data = arrows_df,
               aes(x = 0, y = 0, xend = PCoA1, yend = PCoA2),
               arrow = arrow(length = unit(0.3, "cm")),
               color = "black",
               inherit.aes = FALSE) +  
  geom_text(data = arrows_df,
            aes(x = PCoA1, y = PCoA2, label = Variable),
            size = 4, hjust = 0.5, vjust = -0.7,
            inherit.aes = FALSE) +   
  scale_color_manual(values = cols) +
  theme_minimal() +
  labs(x = "PCoA 1", y = "PCoA 2") +
  theme(legend.position = "none")
#ggsave("PCoA.png")

####TOPSOIL
bray_dist_top <- phyloseq::distance(ps_hel_topsoil, method="bray")
# Perform PCoA
pcoa_top <- cmdscale(bray_dist_top, k = 2)  # k = dimensions
# Create a data frame for the PCoA results
pcoa_top_df <- data.frame(PCoA1 = pcoa_top[, 1],
                      PCoA2 = pcoa_top[, 2],
                      Sample = rownames(pcoa_top))
meta_data_top <- meta_data_top %>%
  mutate(Sample = rownames(meta_data_top))

# Merge with metadata
pcoa_top_df <- merge(pcoa_top_df, meta_data_top, by = "Sample")

# Add envfit:
env2_TOP <- ps_hel_meta_data %>%
  filter(DEPTH =="topsoil") %>%
  dplyr::select(all_of(chem_vars)) 

# Run envfit using the distance matrix and the environmental variables
envfit_top <- vegan::envfit(pcoa_top, env2_TOP, permutations = 999)

arrows_df <- as.data.frame(scores(envfit_top, display = "vectors")[envfit_top$vectors$pvals < 0.01, ])
arrows_df$Variable <- rownames(arrows_df)
mult <- 1.5  # para ajustar la longitud de las flechas al grÃ¡fico
arrows_df <- arrows_df %>%
  dplyr::mutate(PCoA1 = Dim1 * mult,
                PCoA2 = Dim2 * mult)

# Plot PCoA
pcoa_top_df %>%
  ggplot(aes(x = PCoA1, y = PCoA2, color = TRAT2)) +
  geom_segment(data = arrows_df,
               aes(x = 0, y = 0, xend = PCoA1, yend = PCoA2), # xend and yend *0.5 to reduce arrow lenght
               arrow = arrow(length = unit(0.2, "cm")),
               size =1, color = "grey",
               inherit.aes = FALSE) +  
  geom_text(data = arrows_df,
            aes(x = PCoA1, y = PCoA2, label = Variable),
            size = 4, hjust = 0.5, vjust = -0.7,
            inherit.aes = FALSE) +  
  geom_point(size = 5, alpha=.6) +
  scale_color_manual(values = cols) +
  theme_minimal() +
  labs(x = "PCoA 1", y = "PCoA 2") +
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),  # Change this value as needed
    axis.text.y = element_text(size = 14),
    legend.position = "none") +
  scale_x_reverse()
#ggsave("PCoA_top.png")

####SUBSOIL
bray_dist_sub <- phyloseq::distance(ps_hel_subsoil, method="bray")
# Perform PCoA
pcoa_sub <- cmdscale(bray_dist_sub, k = 2)  # k = dimensions
# Create a data frame for the PCoA results
pcoa_sub_df <- data.frame(PCoA1 = pcoa_sub[, 1],
                      PCoA2 = pcoa_sub[, 2],
                      Sample = rownames(pcoa_sub))
meta_data_sub <- meta_data_sub %>%
  mutate(Sample = rownames(meta_data_sub))

# Merge with metadata
pcoa_sub_df <- merge(pcoa_sub_df, meta_data_sub, by = "Sample")

# Add envfit:
env2_SUB <- ps_hel_meta_data %>%
  filter(DEPTH =="subsoil") %>%
  dplyr::select(all_of(chem_vars)) 

# Run envfit using the distance matrix and the environmental variables
envfit_sub <- vegan::envfit(pcoa_sub, env2_SUB, permutations = 999)

arrows_df <- as.data.frame(scores(envfit_sub, display = "vectors")[envfit_top$vectors$pvals < 0.01, ])
arrows_df$Variable <- rownames(arrows_df)
mult <- 1.5  # para ajustar la longitud de las flechas al grÃ¡fico
arrows_df <- arrows_df %>%
  dplyr::mutate(PCoA1 = Dim1 * mult,
                PCoA2 = Dim2 * mult)

# Plot PCoA
pcoa_sub_df %>%
  ggplot(aes(x = PCoA1, y = PCoA2, color = TRAT2)) +
  geom_segment(data = arrows_df,
               aes(x = 0, y = 0, xend = PCoA1, yend = PCoA2), # xend and yend *0.5 to reduce arrow lenght
               arrow = arrow(length = unit(0.2, "cm")),
               size =1, color = "grey",
               inherit.aes = FALSE) +  
  geom_text(data = arrows_df,
            aes(x = PCoA1, y = PCoA2, label = Variable),
            size = 4, hjust = 0.5, vjust = -0.7,
            inherit.aes = FALSE) +  
  geom_point(size = 5, alpha=.6) +
  scale_color_manual(values = cols) +
  theme_minimal() +
  labs(x = "PCoA 1", y = "PCoA 2") +
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),  # Change this value as needed
    axis.text.y = element_text(size = 14),
    legend.position = "none") +
  scale_x_reverse()
#ggsave("PCoA_sub.png")
```

```{r PCoA as linear predictor}
#TOPSOIL
lm_pcoa_top <- lm(PCoA1 ~ TRAT2, data = pcoa_top_df)
summary(lm_pcoa_top)

ggplot(pcoa_top_df, aes(x = TRAT2, y = PCoA1, color=TRAT2, fill = TRAT2)) +
  geom_boxplot(alpha = 0.3) +
  geom_jitter(width = 0.1, aes(color = TRAT2)) +
  scale_fill_manual(values = cols) +
  scale_color_manual(values = cols) +
  theme_minimal()+
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),  # Change this value as needed
    axis.text.y = element_text(size = 14),
    legend.position = "none"
  )
#ggsave("lm_pcoa_top.png")

anova(lm(PCoA1 ~ TRAT2, data = pcoa_top_df))
Tuckeys <- TukeyHSD(aov(PCoA1 ~ TRAT2, data = pcoa_top_df))
cld <- multcompLetters4(lm_pcoa_top, Tuckeys)
print(cld)

#SUBSOIL
lm_pcoa_sub <- lm(PCoA1 ~ TRAT2, data = pcoa_sub_df)
summary(lm_pcoa_sub)

ggplot(lm_pcoa_sub, aes(x = TRAT2, y = PCoA1, color=TRAT2, fill = TRAT2)) +
  geom_boxplot(alpha = 0.3) +
  geom_jitter(width = 0.1, aes(color = TRAT2)) +
  scale_fill_manual(values = cols) +
  scale_color_manual(values = cols) +
  theme_minimal()+
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),  # Change this value as needed
    axis.text.y = element_text(size = 14),
    legend.position = "none"
  )
#ggsave("lm_pcoa_sub.png")

anova(lm(PCoA1 ~ TRAT2, data = pcoa_sub_df))
Tuckeys <- TukeyHSD(aov(PCoA1 ~ TRAT2, data = pcoa_sub_df))
cld <- multcompLetters4(lm_pcoa_sub, Tuckeys)
print(cld)
```

```{r Combined plots for PCoA}
# PCoA SUBSOIL
p3 <- pcoa_top_df %>%
  ggplot(aes(x = PCoA1, y = PCoA2, color = TRAT2)) +
  geom_segment(data = arrows_df,
               aes(x = 0, y = 0, xend = PCoA1*0.5, yend = PCoA2*0.5),
               arrow = arrow(length = unit(0.2, "cm")),
               size = 1, color = "grey",
               inherit.aes = FALSE) +
  geom_point(size = 5, alpha = .6) +
  scale_color_manual(values = cols) +
  theme_minimal() +
  labs(x = "PCoA 1", y = "PCoA 2") +
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    legend.position = "none"
  ) +
  scale_x_reverse()

# Boxplot
p4 <- ggplot(lm_pcoa_top, aes(x = TRAT2, y = PCoA1, color = TRAT2, fill = TRAT2)) +
  geom_boxplot(alpha = 0.2) +
  geom_jitter(width = 0.1, size = 2) +
  scale_fill_manual(values = cols) +
  scale_color_manual(values = cols) +
  theme_minimal() +
  labs(x = "Treatment", y = "PCoA 1") +
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    legend.position = "none"
  )
combined_plot <- p3 + p4 + plot_layout(ncol = 2, widths = c(2, 1))
combined_plot
#ggsave("combined_topsoil.png")

# PCoA SUBSOIL
p1 <- pcoa_sub_df %>%
  ggplot(aes(x = PCoA1, y = PCoA2, color = TRAT2)) +
  geom_segment(data = arrows_df,
               aes(x = 0, y = 0, xend = PCoA1*0.5, yend = PCoA2*0.5),
               arrow = arrow(length = unit(0.2, "cm")),
               size = 1, color = "grey",
               inherit.aes = FALSE) +
  geom_point(size = 5, alpha = .6) +
  scale_color_manual(values = cols) +
  theme_minimal() +
  labs(x = "PCoA 1", y = "PCoA 2") +
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    legend.position = "none"
  ) +
  scale_x_reverse()

# Boxplot
p2 <- ggplot(lm_pcoa_sub, aes(x = TRAT2, y = PCoA1, color = TRAT2, fill = TRAT2)) +
  geom_boxplot(alpha = 0.2) +
  geom_jitter(width = 0.1, size = 2) +
  scale_fill_manual(values = cols) +
  scale_color_manual(values = cols) +
  theme_minimal() +
  labs(x = "Treatment", y = "PCoA 1") +
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    legend.position = "none"
  )
combined_plot <- p1 + p2 + plot_layout(ncol = 2, widths = c(2, 1))
combined_plot
#ggsave("combined_subsoil.png")
```

```{r PERMANOVA}

permanova <- vegan::adonis2(bray_dist ~ TRAT2 * DEPTH, permutations = 999, data = ps_hel_meta_data)
permanova
# Check dispersion
disp <- betadisper(bray_dist, ps_hel_meta_data$TRAT2)
anova(disp)

####TOPSOIL
permanova_top <- vegan::adonis2(bray_dist_top ~ TRAT2, permutations = 999, data = meta_data_top)
permanova_top

meta_data_top <- ps_hel_meta_data %>% dplyr::filter(DEPTH=="topsoil")

disp_top <- betadisper(bray_dist_top, meta_data_top$TRAT2)
anova(disp_top)

####SUBSOIL
permanova_sub <- vegan::adonis2(bray_dist_sub ~ TRAT2, permutations = 999, data = meta_data_sub)
permanova_sub

meta_data_sub <- ps_hel_meta_data %>% dplyr::filter(DEPTH=="subsoil")

disp_sub <- betadisper(bray_dist_sub, meta_data_top$TRAT2)
anova(disp_sub)

#Model R2 = the proportion of total variation in the Bray-Curtis distances that is explained by TRAT2
#Residual R2: This is the proportion of variation not explained by TRAT2
```

```{r RandomForest} 
# It's a regression formula. We should have a data.frame were the first column is the response variable and the remaining columns are the predictor variables
RF_indicators <- as.data.frame(alpha_fungi[, c(2, 35:39, 41:45, 48, 50, 51,53, 54:56, 57,63,68:70,74,78,92,94:96)])
RF_plot <- as.data.frame(alpha_fungi[, c(2, 17:34)])
str(RF_indicators)

summary(RF_indicators$observed)

data <- RF_indicators
data <- data %>% #This implies that column ncol(data) is the response variable
  dplyr::select(ncol(data), everything())
#Model training
#set.seed(12)
RFfit<- randomForest(observed ~ ., data=data, ntree=300, importance=T) # necesario importance=T para guardar el valor de las importancias
summary(RFfit)
#Error plot
plot(RFfit)

#Variable importance
importance_matrix <- importance(RFfit)
incMSE_values <- importance_matrix[, "%IncMSE"]

#png("RFfit_plot.png", width = 800, height = 600)
varImpPlot(RFfit)
#dev.off()
#incMSE indica el error por permutacion
#incNodePurity indica la capacidad de esa variable para separar nodos

# Convertir a dataframe
incMSE_df <- data.frame(Variable = rownames(importance_matrix),
                        IncMSE = importance_matrix[, "%IncMSE"])
# Ordenar las variables de mayor a menor incremento en el MSE
incMSE_df <- incMSE_df[order(incMSE_df$IncMSE, decreasing = TRUE), ]
# Ver los resultados
print(incMSE_df)

ggplot(incMSE_df, aes(x = reorder(Variable, IncMSE), y = IncMSE, fill = IncMSE)) +
  geom_col() +
  #scale_fill_viridis(option="D",direction = -1) +
  coord_flip() +
  theme_minimal() +
  theme(
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black"),
    axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    strip.placement = "outside",  # Place strip labels outside the plot
    strip.background = element_blank()  # Remove strip background
  )
#ggsave("RandomForest_observedrichness.png", height = 5, width = 6, dpi = 300)
```

*We've run all these analysis with the raw data and the Hellinger transformed data and results are practically the same*

The *IndicSpecies Analysis* tells us:
-Which ASVs (or taxa) are statistically associated with particular treatment groups (UB, LS, HS1, HS2).
-A value = specificity (how exclusive the ASV is to a group),
-B value = fidelity (how frequent the ASV is in that group)
*Good for identifying indicator taxa for specific conditions* (e.g., taxa that define unburned soils vs high-severity fire soils)

```{r Indicator Species Analysis}
## Species are chosen as indicators if they:
# 1) Reflect the biotic or abiotic state of the environment;
# 2) Provide evidence for the impacts of environmental change; or
# 3) Predict the diversity of other species, taxa or communities within an area.

#Indicator species are often determined using an analysis of the relationship between the species occurrence or abundance values from a set of sampled sites and the classification of the same sites into site groups, which may represent habitat types, community types, disturbance states, etc. Thus, there are two data elements in an indicator species analysis: 
###############################
# (1) the community data matrix --> with sites in rows and species in columns
# (2) the vector that describes the classification of sites into groups --> a vector containing the classification of the sites into groups
ps1_genus_meta_data <- data.frame(sample_data(ps1_genus))

df <-  as.data.frame((otu_table(ps1_genus))) #community data (abundance matrix)
head(df)

# Transpose the table
asv_transposed <- as.data.frame(t(df))
head(asv_transposed)
asv_transposed <- as.matrix(asv_transposed)

#Now we need to code the vector of site groups. We want to do 4 groups: UB, LS, HS1, HS2
head(ps1_genus_meta_data$TRAT2)
groups4 <- as.factor(ps1_genus_meta_data$TRAT2)

#Indicator species analysis
indval <- indicspecies::multipatt(asv_transposed, groups4, 
                    control = how(nperm=999)) # El argumento control = how(nperm=999) permite elegir el nÃºmero de permutaciones aleatorias requeridas para la prueba de permutaciÃ³n (este nÃºmero afecta la precisiÃ³n del valor p). La funciÃ³n how() del paquete permute permite definir diseÃ±os permutacionales mÃ¡s complejos
summary(indval) 

indval_results <- indval$sign
#The first columns indicate (with ones and zeroes) which site groups were included in the combination preferred by the species. Then, the column indicates the index of the site group combination. The remaining two columns are the association statistic and the p-value of the permutational test.
indval_df <- data.frame(ASV = rownames(indval_results), indval_results)
df_tax <-  as.data.frame((tax_table(ps1_genus)))
df_tax <- df_tax %>% 
  tibble::rownames_to_column(var = "ASV")
indicator_with_taxonomy4 <- merge(indval_df, df_tax, by = "ASV")

# AÃ±adir indvalcomp = TRUE" a la funciÃ³n anterior nos brinda informaciÃ³n adicional sobre por quÃ© las especies pueden usarse como indicadores:
summary(indval, indvalcomp=TRUE)
# A value will tell us if the species is a good indicator of that group (1 = the best indicator)
# B value will tell us if the species appears in all the plots of the group 

# Al aÃ±adir alfa = 1, decimos que queremos mostrar el grupo al que estÃ¡ asociada cada especie, independientemente de si la asociaciÃ³n es significativa o no. (por defecto viene establecido alpha 0,05, que se corresponde al valor de significancia que  tiene en cuenta el anÃ¡lisis de permutaciÃ³n)  
summary(indval, indvalcomp=TRUE, alpha=1) 

#######################################################################
#Now we want to do 8 groups: UB_topsoil, UB_subsoil, LS_topsoil, LS_subsoil, HS1_topsoil, HS1_subsoil, HS2_topsoil, HS2_subsoil
head(ps_TOP_meta_data$TRAT2_DEPTH)
groups8 <- as.factor(ps_TOP_meta_data$TRAT2_DEPTH)

#Indicator species analysis
indval8 <- indicspecies::multipatt(asv_transposed, groups8, 
                    control = how(nperm=999))
summary(indval8)

indval8_results <- indval8$sign
#The first columns of indicate (with ones and zeroes) which site groups were included in the combination preferred by the species. Then, the column indicates the index of the site group combination. The remaining two columns are the association statistic and the p-value of the permutational test.
indval8_df <- data.frame(ASV = rownames(indval8_results), indval8_results)

indicator_with_taxonomy8 <- merge(indval8_df, df_tax, by = "ASV")

############################## plot ####################################
###BUBBLE
# Convert wide format to long format
plot_indval_data4 <- indicator_with_taxonomy4 %>%
  pivot_longer(cols = c(s.UB, s.LS, s.HS1, s.HS2), 
               names_to = "TRAT2", 
               values_to = "Presence") %>%
  
  dplyr::filter(p.value < 0.05) #Only significant taxa

plot_indval_data8 <- indicator_with_taxonomy8 %>%
  pivot_longer(cols = c(s.UB_topsoil, s.UB_subsoil, s.LS_topsoil, s.LS_subsoil, 
                        s.HS1_topsoil, s.HS1_subsoil, s.HS2_topsoil, s.HS2_subsoil), 
               names_to = "TRAT2", 
               values_to = "Presence") %>%
  
  dplyr::filter(p.value < 0.05) #Only significant taxa

```

```{r Species Indicator Analysis: plot}
indicator_with_taxonomy4$Genus_clean <- gsub("^g__", "", indicator_with_taxonomy4$Genus)

filtered_df <- indicator_with_taxonomy4 %>%
  filter(index %in% c(1, 10, 14)) #We just want these groups

indval_traits <- filtered_df %>%
  left_join(fungal_traits, by = c("Genus_clean" = "Genus"))

# Convert wide format to long format
plot_indval_data4 <- indval_traits %>%
  pivot_longer(cols = c(s.UB, s.LS, s.HS1, s.HS2), 
               names_to = "TRAT2", 
               values_to = "Presence") %>%
    dplyr::filter(p.value < 0.05) #Only significant taxa

plot_indval_data4$TRAT2 <- ordered(plot_indval_data4$TRAT2, levels = c("s.UB", "s.LS", "s.HS1", "s.HS2"))

#HEATMAP
ggplot(plot_indval_data4, aes(x = TRAT2, y = trophic_mode_fg, fill = Presence)) +
  geom_tile(color = "white") +  # Add white grid lines for clarity
  scale_fill_gradient(low = "white", high = "blue") +  # Adjust color scale
  theme_minimal() +
  labs(fill = "Indicator Value") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels
#ggsave("indicspecies_top10order_heatmap.png")

# DOTS
ggplot(plot_indval_data4, aes(x = TRAT2, y = guild_fg)) +
  geom_point(
    aes(fill = as.factor(Presence)), 
    shape = 21,      # Allows fill + border
    color = "black", # Border color
    alpha=0.9,
    size = 5
    
  ) +
  scale_fill_manual(values = c("0" = "white", "1" = "lightblue")) +
  theme_minimal() +
  labs(fill = "Presence") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.grid = element_blank()
  )
#ggsave("indicspecies_guild.png")
```

*Similarity Percentages Analysis* (SIMPER)
- It decomposes Bray-Curtis dissimilarities between groups into contributions from individual taxa.
- Good for identifying *which taxa drive differences in community composition* between pairs of groups.

```{r SIMPER}
simper_results <- simper(asv_transposed, groups4, permutations = 999)
summary(simper_results)
```

The main conclusion we can get out of this analysis is that there's a significant impact of wildfire severity on the topsoil while in the subsoil the difference is only between burnt vs unburnt

```{r Redundancy Analysis (RDA)}
otu_mat <- t(data.frame(otu_table(ps1)))
metadata <- data.frame(sample_data(ps1))

vare.cap <- vegan::capscale(otu_mat ~ TRAT2*DEPTH , metadata, dist="hellinger") ## dist = bray, hellinger â€¦
#Use 'Condition()' to account for random factors (e.g., block or plot), if needed
summary(vare.cap)
#Partitioning of squared Hellinger distance:
#              Inertia Proportion
#Total          56.222    1.00000
#Conditioned     4.299    0.07646     #7.6% explained by RANDOM factor
#Constrained     5.037    0.08958     #8.9% explained by TRAT2 , DEPTH and their interaction
#Unconstrained  46.887    0.83396     #83% without explanation

permutest(vare.cap, permutations=how(nperm=9999)) #permutation test partial significant
#Model: vegan::capscale(formula = otu_mat ~ TRAT2 * DEPTH + Condition(SITE), data = metadata, distance = "hellinger")
#Permutation test for all constrained eigenvalues
#         Df Inertia      F Pr(>F)
#Model     7   5.037 1.0895 0.2056   #Model significance
#Residual 71  46.887 


anova(vare.cap, by="term")#1e-04 *
anova.cca(vare.cap, permutations = 10000, by = "axis") #permutation test to see axis significance: RDA1 significant
anova.cca(vare.cap, permutations = 10000, by = "terms") #permutation test to see axis significance: RDA1 significant

RsquareAdj(vare.cap) #total variance explained by RDA

#extraer puntos
scrs <- scores(vare.cap, display = "sites")

rownames (scrs) == rownames (meta_data_pc1)

RDA=cbind(metadata,scrs) #puntos del RDA por tratamiento para graficar

 

ggplot() +
  geom_point(data= RDA, aes(CAP1, CAP2, color=TRAT2, shape=DEPTH), size=2, alpha=0.7,stroke=1.5) +
  stat_ellipse(data= RDA, aes(CAP1, CAP2, color=TRAT2), geom="path", size=0.5, level=0.5) +
  #scale_colour_manual(values=cols) +
  #scale_shape_manual(values = c("conventional" = 16, "reduced" = 1)) +  # Define los sÃ­mbolos
  # theme_bw() +
  theme(axis.title=element_text(size=8)) +
  theme(axis.ticks=element_line()) +
  theme_cowplot(10) +
  theme(axis.text=element_text(size=8, colour="black")) +
  theme(legend.position = "none") +
  ylab("db-RDA2 (48.2%)") +
  xlab("db-RDA1 (18.9%)") +

  #my_theme

 

#aÃ±adir variables ambientales con envfit para graficarlas posteriormente en el RDA plot

CAP_env <- envfit(vare.cap, metadata[,c(5,6,10:27,33,66,72:76)], permutation=9999,display = "sites", strata=NULL,na.rm = TRUE,w = NULL)

 

vectors_CAP <- scores(CAP_env,display=c("vectors"))

vectors_CAP <- as.data.frame(vectors_CAP)

vectors_CAP $variable <- rownames(vectors_CAP)

 

vec.p <- cbind(vectors_CAP, pval = CAP_env $vectors$pvals) #add pvalues to dataframe so you can select variables which are significant

r_squared <- CAP_env $vectors$r

vec.p <- cbind(vec.p, r_squared = r_squared)

vec.sig <- subset(vec.p, pval<=0.05) #sub
```

```{r}
otu <- as.data.frame(otu_table(ps_sub))
  if (taxa_are_rows(ps_sub)) otu <- t(otu)
  comm_dist <- vegdist(otu, method = "bray")
 
  env <- as.data.frame(sample_data(ps_sub))
  env_manual <- env[, variables_manual, drop = FALSE]
  env_manual <- env_manual[, sapply(env_manual, function(x) is.numeric(x) && all(is.finite(x))), drop = FALSE]
 
  if (ncol(env_manual) == 0) {
    message("No hay variables numÃ©ricas vÃ¡lidas para ", parque)
    next
  }
 
  env_scaled <- scale(env_manual)

  # PERMANOVA
  adonis_res <- adonis2(comm_dist ~ ., data = as.data.frame(env_scaled), permutations = 999)
  print(adonis_res)
  write.csv(as.data.frame(adonis_res), file = file.path(output_dir, paste0("permanova_", parque, ".csv")))

  # db-RDA
  rda_model <- capscale(comm_dist ~ ., data = as.data.frame(env_scaled))
  print(summary(rda_model))

  # ANOVA
  anova_global <- anova(rda_model, permutations = 999)
  anova_margin <- anova(rda_model, by = "margin", permutations = 999)
  print(anova_global)
  print(anova_margin)

  write.csv(as.data.frame(anova_global), file = file.path(output_dir, paste0("anova_global_", parque, ".csv")))
  write.csv(as.data.frame(anova_margin), file = file.path(output_dir, paste0("anova_variables_", parque, ".csv")))

  # Coordenadas
  site_scores <- scores(rda_model, display = "sites")
  env_scores <- scores(rda_model, display = "bp")

  write.csv(site_scores, file = file.path(output_dir, paste0("sitescores_", parque, ".csv")))
  write.csv(env_scores, file = file.path(output_dir, paste0("envvectors_", parque, ".csv")))

  # -------------------------------------
  # VISUALIZACIÃ“N
  # -------------------------------------

  # Extraer orientaciÃ³n y asegurar orden correcto
  site_df <- as.data.frame(site_scores)
  site_df$COTA <- sample_data(ps_sub)$COTA[match(rownames(site_df), rownames(sample_data(ps_sub)))]
  site_df$ORIENTACION <- sample_data(ps_sub)$ORIENTACION_NORTE_SUR[match(rownames(site_df), rownames(sample_data(ps_sub)))]

  # ClasificaciÃ³n de variables por significancia
  anova_margin_df <- as.data.frame(anova_margin)
  anova_margin_df$Variable <- rownames(anova_margin_df)
  anova_margin_df$Significancia <- dplyr::case_when(
    anova_margin_df$`Pr(>F)` < 0.05 ~ "Significativa",
    anova_margin_df$`Pr(>F)` < 0.1  ~ "Marginal",
    TRUE ~ "No"
  )

  vars_sig   <- anova_margin_df %>% filter(Significancia == "Significativa") %>% pull(Variable)
  vars_marg  <- anova_margin_df %>% filter(Significancia == "Marginal") %>% pull(Variable)

  env_sig <- env_scores[vars_sig, , drop = FALSE]
  env_marg <- env_scores[vars_marg, , drop = FALSE]

  env_df <- bind_rows(
    as.data.frame(env_sig)  %>% mutate(tipo = "solid", Variable = rownames(env_sig)),
    as.data.frame(env_marg) %>% mutate(tipo = "dashed", Variable = rownames(env_marg))
  )

  # GrÃ¡fico db-RDA
  plot_rda <- ggplot() +
    geom_point(data = site_df,
               aes(x = CAP1, y = CAP2, color = COTA, shape = ORIENTACION),
               size = 3, alpha = 0.8) +
    geom_segment(data = env_df,
                 aes(x = 0, y = 0, xend = CAP1, yend = CAP2, linetype = tipo),
                 arrow = arrow(length = unit(0.2, "cm")), color = "black") +
    geom_text(data = env_df,
              aes(x = CAP1, y = CAP2, label = Variable),
              hjust = 1.2, vjust = 1.2, size = 4) +
    scale_linetype_manual(values = c("solid" = "solid", "dashed" = "dashed"),
                          labels = c("solid" = "p < 0.05", "dashed" = "0.05 â‰¤ p < 0.1"),
                          name = "Variables ambientales") +
    labs(title = paste("db-RDA -", parque),
         x = "db-RDA 1", y = "db-RDA 2",
         color = "COTA", shape = "OrientaciÃ³n") +
    theme_minimal() +
    theme(legend.position = "right")

  print(plot_rda)

```


```{r Ergosterol}
ggplot(alpha_fungi, aes(x = pc1, y = Ergosterol_Âµg_g, color = TRAT2)) +
  geom_point(size = 4, alpha = 0.7) +
  scale_color_manual(values = cols) + 
  geom_smooth(method = "lm", se = TRUE, color = "black", linetype = "dashed") +
  theme_minimal()+
    theme(
    #legend.position = "none",
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "black"),
    axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    strip.placement = "outside",  # Place strip labels outside the plot
    strip.background = element_blank()  # Remove strip background
  ) +
  scale_x_reverse()
#ggsave("ergosterol_pc1.png")
```


############################ EXTRA ##############################################

```{r Sankey Diagram}
## We need two dataframes:
# Nodes: with a number for each node we want to represent (ins and outs) --> e.g.: 1-UB, 2-LS, 3-HS1... Two columns
# Links: Three columns. Source (from 0 to nÂº nodes - 1), Target and Value
h <- ps3meltrel %>%
  dplyr::group_by(TRAT2, DEPTH, Phylum) %>%
  dplyr::summarise(total_abundance = sum(Abundance, na.rm = TRUE))

#write.csv(h, "ps4melted.csv") --> we make the codification in Excel

links <- as.data.frame(read_excel("C:/Users/amara.santiesteban/OneDrive - UAM/Escritorio/Sankey_trial.xlsx", sheet = "Links"))
nodes <- as.data.frame(read_excel("C:/Users/amara.santiesteban/OneDrive - UAM/Escritorio/Sankey_trial.xlsx", sheet = "Nodes"))

sankey <- networkD3::sankeyNetwork(Links = links, Nodes = nodes, Source = "source",
               Target = "target", Value = "Value", NodeID = "names",  iterations = 0) #Setting "iterations = 0" will disable the algorithm which automatically determines the node placement, and the nodes will be placed in the order that they appear in the Nodes dataframe.

pdf_file <- "sankey_diagram.pdf"
webshot2::webshot(html_file, file = pdf_file)
```

```{r Comparation with BLAST}
#top5_UNITE22 <- as.data.frame(read_xlsx("Libro1.xlsx", col_names=TRUE, sheet = 1))
head(top5_UNITE22)

ggplot(top5_UNITE22, aes(fill=Family2, y=Count, x=TRAT2)) + 
    geom_bar(position="stack", stat="identity")

top5_UNITE23 <- as.data.frame(read_xlsx("Libro1.xlsx", col_names=TRUE, sheet = 2))
```

```{r Rarefaction curves}
devtools::install_github("gauravsk/ranacapa")

plot <- ranacapa::ggrare(ps1, step = 100, label = NULL, color = "ID", se = TRUE)
#(plot <- plot + facet_wrap(~TRAT2))
#ggsave("rarefactionps3.png")
```

*NMDS (Non-Metric Multidimensional Scaling)*
-A non-linear non-metric ordination technique.
-Ranks distances rather than preserving actual values.
-Only assumes that rank order of distances is meaningful.
-Axes are arbitrary and do not have intrinsic meaning.
-The distance between points is only relative (ordinal).
Best use case: When working with non-Euclidean distances (e.g., Bray-Curtis) and non-normal ecological data.
ðŸŸ¢ Advantages:
More robust for ecological data with non-linear patterns.
Works well with Bray-Curtis and other ecological dissimilarity metrics.
ðŸ”´ Disadvantages:
Cannot directly compare axes across different NMDS runs.
Stress values must be interpreted to determine fit (stress < 0.2 is acceptable, < 0.1 is ideal).

- We can use the vegdist function to calculate our distance matrix if we need it for further analysis, but the *metaMDS function* will automatically calculate the distance matrix based on the type of distance we choose.

```{r NMDS}
OTU.ord <- ordinate(ps_hel, "NMDS", "bray")

ps_hel_meta_data <- data.frame(sample_data(ps_hel))  # Get the metadata from the ps object

ps_hel_meta_data <- ps_hel_meta_data %>% 
  mutate(Fe_g_Kg = Fe_mg_Kg_105/1000,
         Ergosterol_Âµg_g = Ergosterol_ng_g_105/1000,
         C_N_ratio = Ctotal_g_100g_105/Ntotal_g_100g_105)

plot_vars <- c("Slope_per","Organic_layer_mean_depth_cm", "Charred_organic_layer_mean_cm", "Green_per", "Brown_per", "Consumed_per", "Grass_per", "Old_litter_per", "New_litter_per", "Mineral_per", "Ash_per", "Stone_per")

env_df <- ps_hel_meta_data %>%
  dplyr::select(all_of(plot_vars))

chem_vars <- c("pH", "EC", "densidad_revisada_25", "C_N_ratio",
              "A_phosphorus_ppm_105", "Phosphatase_activity_Âµmol_g_h", 
              "NH4_mgL_105","NO3_mgL_105", "Ergosterol_Âµg_g", 
              "Ntotal_g_100g_105", "Ctotal_g_100g_105", "Carbon_stock_Mg_ha",
              "Nitrogen_stock_Mg_ha", "SOM_per_corregido", "MBC_mg_kg_105",
              "Nematode_g_105", "Ca_g_kg", "K_g_kg","Mg_g_kg", "Na_g_kg", 
              "Al_g_kg", "Fe_g_Kg","Zn_mg_Kg_105", "Cu_mg_Kg_105")

env2_df <- ps_hel_meta_data %>%
  dplyr::select(all_of(chem_vars))

envfit_results  <- vegan::envfit(OTU.ord, env2_df, permutations = 999, na.rm = TRUE)

env_coord_sig0 <- scores(envfit_results, "vectors")[envfit_results$vectors$pvals < 0.01, ] #Only the significant variables

# Plot the ordination and add the environmental vectors
plot_ordination(ps_hel, OTU.ord, color = "TRAT2") +
  geom_point(size = 4) +  # Adjust point size
  geom_polygon(alpha = 0.2) +
  scale_fill_manual(values = cols) + 
  scale_colour_manual(values = cols) +
  geom_segment(data = as.data.frame(env_coord_sig0), 
               aes(x = 0, y = 0, xend = NMDS1, yend = NMDS2),
               arrow = arrow(type = "open", length = unit(0.2, "inches")), 
               size = 1, color = "blue") +
  geom_text(data = as.data.frame(env_coord_sig0), 
            aes(x = NMDS1, y = NMDS2, label = rownames(env_coord_sig0)),
            size = 3, color = "blue") +
  theme_minimal()
#ggsave("nmds_hel.png")

#####TOPSOIL
ps_hel_topsoil <- subset_samples(ps_hel, DEPTH == "topsoil")

# Perform ordination for topsoil samples
OTU.ord_topsoil <- ordinate(ps_hel_topsoil, method = "NMDS", distance = "bray")

#Add environmental data
env_df_TOP <- ps_hel_meta_data %>%
  filter(DEPTH == "topsoil") %>%
  dplyr::select(all_of(chem_vars))

envfit_TOP  <- vegan::envfit(OTU.ord_topsoil, env_df_TOP, permutations = 999, na.rm = TRUE)

envfit_TOP_sig <- scores(envfit_TOP, "vectors")[envfit_TOP$vectors$pvals < 0.01, ] #Only the significant variables

# Plot the ordination and add the environmental vectors
plot_ordination(ps_hel, OTU.ord_topsoil, color = "TRAT2") +
  geom_point(size = 4) +  # Adjust point size
  geom_polygon(alpha = 0.1) +
  scale_fill_manual(values = cols) + 
  scale_colour_manual(values = cols) +
  geom_segment(data = as.data.frame(envfit_TOP_sig), 
               aes(x = 0, y = 0, xend = NMDS1, yend = NMDS2),
               arrow = arrow(type = "open", length=unit(0.06, "npc")), 
               size = 1, color = "#737373") +
  geom_text(data = as.data.frame(envfit_TOP_sig), 
            aes(x = NMDS1, y = NMDS2, label = rownames(envfit_TOP_sig)),
            size = 3, color = "#737373") +
  scale_x_reverse()+
    theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),  # Change this value as needed
    axis.text.y = element_text(size = 14),
    legend.position = "none"
  )
#ggsave("nmdsTOP_chem.png")

#####SUBSOIL
ps_hel_subsoil <- subset_samples(ps_hel, DEPTH == "subsoil")

# Perform ordination for subsoil samples
OTU.ord_subsoil <- ordinate(ps_hel_subsoil, method = "NMDS", distance = "bray")

env_df_SUB <- ps_hel_meta_data %>%
  filter(DEPTH == "subsoil") %>%
  dplyr::select(all_of(chem_vars))

envfit_SUB  <- vegan::envfit(OTU.ord_subsoil, env_df_SUB, permutations = 999, na.rm = TRUE)

envfit_SUB_sig <- scores(envfit_SUB, "vectors")[envfit_SUB$vectors$pvals < 0.01, ] #Only the significant variables

# Plot the ordination and add the environmental vectors
plot_ordination(ps_hel, OTU.ord_subsoil, color = "TRAT2") +
  geom_point(size = 4) +  # Adjust point size
  geom_polygon(alpha = 0.2) +
  scale_fill_manual(values = cols) + 
  scale_colour_manual(values = cols) +
  geom_segment(data = as.data.frame(envfit_SUB_sig), 
               aes(x = 0, y = 0, xend = NMDS1, yend = NMDS2),
               arrow = arrow(type = "open", length = unit(0.06, "inches")), 
              size = 1, color = "#737373") +
  geom_text(data = as.data.frame(envfit_SUB_sig), 
            aes(x = NMDS1, y = NMDS2, label = rownames(envfit_SUB_sig)),
            size = 3, color = "#737373") +
  #scale_x_reverse()+
  theme(
    panel.background = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    panel.grid = element_blank(),
    axis.text.x = element_text(size = 14),  # Change this value as needed
    axis.text.y = element_text(size = 14),
    legend.position = "none"
  )
#ggsave("nmdsSUB_chem.png")
```

